OCPEDGE-2280: mutable topology (#2008)
openshift/enhancements: enhancements/topologies/mutable-topology.md

---
title: mutable-topology
authors:
  - "@jeff-roche"
  - "@jaypoulz"
  - "@eggfoobar"
reviewers:
  - "@tjungblu, for cluster-etcd-operator"
  - "@joelspeed, for API and infrastructure config"
  - "@spadgett, for console"
  - "@jerpeter, for OpenShift architecture"
approvers:
  - "@jerpeter, for OpenShift architecture"
api-approvers:
  - "@joelspeed, for API and infrastructure config"
creation-date: 2026-05-11
last-updated: 2026-05-11
tracking-link:
  - https://issues.redhat.com/browse/OCPEDGE-2280
replaces:
  - https://github.com/openshift/enhancements/pull/1905
superseded-by: []
---

# Mutable Topology

## Terms

> **Mutable Topology** — The capability for an OpenShift cluster to transition between topology modes as a Day 2 operation, breaking the existing assumption that topologies are immutable after installation.

Reviewer: ```suggestion
**Mutable Topology** — The capability for an OpenShift cluster to transition between topology modes as a Day 2 operation, removing the existing assumption that topologies are immutable after installation.
```

Author: accepted

Author: Applied in latest push.

**OpenShift Topology Transition Operator (OTTO)** — An optional payload operator responsible for orchestrating topology transitions. OTTO owns the transition DAG, validates preconditions, coordinates with cluster operators, and updates the Infrastructure config once the cluster is ready.

**Control Plane Topology** — The cluster-topology mode describing how control-plane nodes are deployed and managed (SingleReplica, HighlyAvailable, or other supported modes). Control-plane nodes are nodes labeled with `node-role.kubernetes.io/control-plane` or `node-role.kubernetes.io/master`.

**Infrastructure Topology** — The cluster-topology mode describing how infrastructure workloads are distributed (SingleReplica, HighlyAvailable, or other supported modes). When there are no worker nodes, control-plane nodes serve as workers.

**Compact Cluster** — A cluster where control-plane nodes also serve as workers. In the initial SNO-to-HA transition, the target is a 3-node compact cluster with no dedicated worker nodes.

> **Cluster Administrator** — A human user responsible for managing an existing cluster, including Day 2 operations such as topology transitions and node scaling.

Reviewer: ```suggestion
**Cluster Administrator** — An entity responsible for managing an existing cluster, including Day 2 operations such as topology transitions and node scaling.
```
Could be an external orchestrator or agent at this point.

Author: accepted

Author: Applied in latest push.

## Summary

> This enhancement introduces mutable topology — the ability for OpenShift clusters to transition between topology modes as a Day 2 operation. This changes the existing OpenShift assumption that topologies are immutable after installation.

Reviewer: ```suggestion
This enhancement introduces "mutable topology" which is defined as "the ability for OpenShift clusters to transition between topology modes as a Day 2 operation". This changes the existing OpenShift assumption that topologies are immutable after installation.
```

Author: accepted

Author: Applied in latest push.

> A new optional payload operator, the OpenShift Topology Transition Operator (OTTO), will orchestrate transitions.

Reviewer: Is it technically feasible to do this in the existing core operators without creating a new operator?

Author: Technically yes, but it's likely not the best choice due to factors covered later in the proposal and that I mention in other discussion threads (resolving this one in favor of those other threads)

where operators dynamically react to node count changes.
That approach required updating every core operator to handle dynamic topology shifts
and introduced a new topology enum value that all operators had to understand.
It also coupled topology behavior to node count, making operator logic more complex.

Mutable topology takes a different approach: instead of adding a new topology mode that operators must interpret,
transitions are orchestrated by a dedicated operator that coordinates the sequencing, validates preconditions,
and updates the Infrastructure config only when the cluster is ready for the new mode.
Operators continue to react to the same fixed topology values they already understand.
This keeps operator logic simple and concentrates transition complexity in a single component.

### User Stories

* As a cluster administrator running Single Node OpenShift (SNO) at an edge location, I want to add control-plane nodes to my cluster to achieve high availability so that I can handle node failures without service disruption as workloads become more critical.

> * As a solutions architect deploying OpenShift clusters at scale, I want to start with minimal footprint deployments that can grow into highly available clusters so that I can reduce initial costs while maintaining scalability.

Reviewer: ```suggestion
* As a cluster administrator deploying OpenShift clusters at scale, I want to start with minimal footprint deployments that can grow into highly available clusters so that I can reduce initial costs while maintaining scalability.
```
The use of solutions architect seemed strange here as that's generally a sales role.

Author: accepted

Author: Applied in latest push.

* As a cluster administrator managing a fleet of edge deployments, I want a supported path to transition my cluster topology so that I don't need to redeploy clusters when my infrastructure requirements change.

* As a platform engineer, I want topology transitions managed by a dedicated operator so that the transition logic is isolated from my operational tooling and I have a clear interface for monitoring transition state.

### Goals

> * Change the OpenShift support guideline of "topologies are immutable" by providing a supported mechanism for topology transitions

Reviewer: ```suggestion
* Officially support topology transitions in OpenShift.
```

Author: accepted

Author: Applied in latest push.

* Provide a topology transition operator (OTTO) that owns the transition DAG and orchestrates transitions safely

> * Provide an `oc adm transition topology` CLI command for interactive transition management

Reviewer: ```suggestion
```
Suggesting that these be dropped because they're more implementation details than goals. We know that the previous EP isn't the path forward but it seems too soon to determine that a new operator is the goal.

Author: I'm going to push back on dropping these, they are the driver and target direction of this new proposal and until/if a different is determined, they are the current path and are goals of this proposal because they are the core drivers and identify where the functionality should live. Breaking into a lot of detail around the operator internals is out of scope but a high level "This is how we think we can do this" is appropriate to give reviewers an idea of the approach we will take.

Reviewer: This isn't a blocker so they can of course stay. For context, the way I view this is the previous proposal was "Let's introduce a new topology" and now the proposal is "Let's make topology mutable". These implementation details are ideas for how to achieve the goal but not goals in and of themselves.

Author: Per our earlier discussion — keeping these as goals. Not a blocker per your follow-up.

- Coupled topology behavior to node count, making operator logic more complex
- Required shared library-go utilities that every operator team needed to adopt
- The `Adaptable` enum value created a paradigm that was fundamentally different from existing fixed topology modes

Mutable topology achieves the same end goal (SNO clusters can grow to HA) with less operator-side complexity. Operators continue to react to the same fixed topology values they already understand. Transition complexity is concentrated in OTTO rather than distributed across all operators.

### CLI-Only Transition Runner

An alternative is to embed all transition logic in the `oc adm transition` command without a dedicated operator.

**Why it was rejected**:
- As more transitions become supported, the CLI would grow unmanageably large
- Long-running transitions require persistent state tracking that a CLI process cannot reliably provide
- The CLI would need direct access to operator internals, violating separation of concerns
- Error recovery and retry logic is better suited to an operator's reconciliation loop

> ### Controller in CVO

Reviewer: Is CVO the only option in the core operators where this might make sense?

Author: I expanded to include some other operators, none of which fit the bill in my opinion. This is an entirely new process and shoehorning it into another operator that wasn't designed for tackling this type of procedure seems irresponsible to me

Reviewer: Which operator handles adding nodes to clusters?

Author: Expanded to include other operators in the Alternatives section. The controller now lives in CCO, which was agreed on in the JoelSpeed/jaypoulz thread.

An alternative is to add transition controllers to the cluster-version-operator (CVO).

**Why it was rejected**:

> - Overloads CVO with responsibilities unrelated to version management

Reviewer: CVO is the Cluster Version Operator. Why is unrelated to version management?

Author: CVO is related to version management, mutable topologies is not. You don't mutate across versions. The version remains the same after transitioning the topology.

Reviewer: Reading this with fresh eyes I realized I misread the statement. For whatever reason, Github isn't allowing me to resolve this thread so ignore this one going forward :smile:

Author: Per your follow-up — ignoring this one.

> - As the transition DAG grows, it would significantly increase CVO complexity

Reviewer: This seems to assume unbounded growth of the DAG. Given that we're talking about topologies, isn't this set pretty fixed so it wouldn't actually increase complexity much?

Author: Each leg of the DAG could have very different requirements and preconditions. While our current set (SNO -> HA) is relatively straightforward graph, things can complex quick (ex: configuration for two node with fencing complicates the logic considerably)

Author: Acknowledged — the transition graph is bounded. Renamed to 'directed graph' throughout.

> - CVO is already a complex operator; adding topology transition logic increases risk

Reviewer: ```suggestion
- Adding topology transition logic increases risk to other CVO functions.
```
Suggesting dropping the preface statement; the complexity of the operator is less informative for this decision than the stability of the operator.

Author: accepted

Author: Applied in latest push.

## Motivation

Cluster demands change over time. Customers who start with Single Node OpenShift (SNO) at edge locations may later require high availability as workloads become more critical. Today, this requires redeploying the cluster — a disruptive operation that involves workload migration, downtime, and operational overhead.

The previous approach to this problem ([Adaptable Topology](https://github.com/openshift/enhancements/pull/1905)) proposed a new topology mode
where operators dynamically react to node count changes.
That approach required updating every core operator to handle dynamic topology shifts
and introduced a new topology enum value that all operators had to understand.
It also coupled topology behavior to node count, making operator logic more complex.

Mutable topology takes a different approach: instead of adding a new topology mode that operators must interpret,
transitions are orchestrated by a dedicated operator that coordinates the sequencing, validates preconditions,
and updates the Infrastructure config only when the cluster is ready for the new mode.
Operators continue to react to the same fixed topology values they already understand.

> This keeps operator logic simple and concentrates transition complexity in a single component.

Reviewer 2: But still this requires checking / visiting all operators, to ensure they react on a topology change on day2. As they might apply their logic only at install time with the (old) assumption that the topology is immutable.

Author: Added a section stating that we need to investigate payload operators for their functionality.

Author: The per-operator topology dependency matrix is now a prerequisite for entering dev preview — see Graduation Criteria. This captures the audit work.

* Scaling down control-plane or worker nodes (scale-down may be addressed in a future enhancement)
* Supporting bidirectional transitions (e.g., HA → SNO) in the initial implementation
* Installing OTTO by default on all clusters

## Proposal

Mutable topology introduces a new optional payload operator and CLI command to enable topology transitions as Day 2 operations.

A dedicated operator is the right vehicle for this because topology transitions are long-running, multi-step orchestration workflows that require persistent state, failure recovery, and coordination across multiple cluster operators. This logic does not belong in CVO — CVO is a critical-path operator where additional surface area increases risk to every cluster, and topology transitions are operationally distinct from version management. It does not belong in the CLI because CLI processes cannot survive disconnects or provide the reconciliation loop needed for automatic recovery from partial failures. A separate operator isolates this complexity, ships with the payload but installs only when needed, and can be tested independently.

The approach has two components:

> 1. **OpenShift Topology Transition Operator (OTTO)** — An operator that ships with the payload but is not installed by default. OTTO owns the transition graph, validates preconditions, orchestrates the transition sequence, and updates the Infrastructure config as the final step.

Reviewer 3: BY which mechanism are you planning to install this only when needed? Once installed, what manages and updates this going forward?

Reviewer 4: I think the context that is missing here is - Jeff and I were discussing ways to make that we aren't adding idling infrastructure on the control-plane nodes to handle transitions until they're ready.

The mechanism to install when needed is the same thing we do with NTO - the CLI tool would create a manifest to deploy a replica for the controller used to run the pre-flights and transitions.

Reviewer 2: we need to make sure that this is handled correctly by offline/airgapped scenarios, e.g. "oc-mirror" so that the required images are available when OTTO is added. 
Another idea: how about making this a cluster capability which is not enabled per default. Cluster Capabilities can be easily added as a day2 operation.

Author: The controller now lives in CCO, gated by the `MutableTopology` feature gate. When the gate is disabled, the controller is not registered and incurs negligible overhead. This addresses the concern about idling infrastructure.

2. **`oc adm transition topology` CLI command** — An interactive command that installs/activates OTTO if needed, guides the administrator through configuring the transition (nodes, certificates, secrets, etc.), and monitors transition status via OTTO's custom resources.

### Workflow Description

#### Transition: SNO to HA Compact (3-Node)

**cluster administrator** is an entity responsible for managing an existing cluster.

**Non-functional constraint**: There is no availability guarantee during topology transitions. Scaling control-plane nodes is an explicit operational action, and administrators should treat it as a maintenance window. The cluster is expected to be fully available before and after the transition, but not necessarily during.

##### Pre-Transition

> 1. The cluster administrator prepares the additional control-plane nodes (hardware, network, OS)

Reviewer 5: Does 'OS' here imply that the user joins the hosts to the cluster as as control plane nodes at this stage? If not, at what stage is that expected to happen?

Reviewer 2: Not sure what this means? Does this mean just prepping the HW is inplace? Or does this mean adding the node as a worker node to the cluster?
That would have the benefit that we can rely on all the existing docs and procedures on how to add a worker node to an existing cluster.

Author: Clarified in the current revision — step 1 now explicitly says 'the kubelet is running on each node and Node objects exist in the Kubernetes API'. The administrator joins nodes before initiating the transition.

2. The cluster administrator runs `oc adm transition topology` to begin the interactive transition flow
3. The CLI checks whether OTTO is installed; if not, it installs/activates it
4. The CLI guides the administrator through providing required configuration:

>    - Node identities and access details for the new control-plane nodes

Reviewer 3: They aren't already Nodes at this point then?

Reviewer 4: I think this is just poor wording. At this stage, the operator is looking to the node API to see if they exist. The user is assumed to have taken steps prior to step 4 to ensure that the nodes are present.

If the nodes are not present yet, OTTO fails the pre-flight checks.

Reviewer 2: IIUC, this means OTTO transforme a worker node into a control plane node. Maybe worth stating this explicitly?

Author: Clarified — step 1 now says 'the kubelet is running on each node and Node objects exist in the Kubernetes API'. The user is assumed to have joined nodes before initiating the transition.

>    - Any certificates or secrets required for the transition

Reviewer 3: What might these secrets/certificates be used for?

Reviewer 4: I am not aware of any certificates, unless Jeff is talked about a pre-flight check for the certificate approvals for the new nodes - e.g. node bootstrapper, client, etc. But those will get sorted by the node controller. If the nodes are ready, we're good to fly.

Author: I misunderstood part of the intent of the CLI, I thought we were going to need to step the user through configuring the nodes, it seems that is not the case and we are assuming the new nodes are pre-configured and we just need to validate that

Author: This was a misunderstanding on my part — no secrets or certificates need to be managed by the transition controller. The pre-transition step assumes nodes are already joined and the node controller has handled certificate approvals.

>    - Load balancing configuration (on `platform: none`, the user manages their own VIPs/DNS)

Reviewer 3: So for the initial implementation, this doesn't actually do anything? On AWS/Azure/GCP, is there manual action required here? Typically when the node is created it is added to the load balancer by whatever creates it, or the CCM depending on platform, I wouldn't expect the transition operator to have needed to know anything about this external infrastructure 

Which platforms will need to update their config?

Reviewer 4: The reason we mention platform is because we only plan to support 1->3 on platform none and possible platform baremetal for the first pass.

To add support for other platforms, the corresponding platform controllers will need to handle stitching up the networks on those nodes. This is why we're looking at baremetal networking and how we can update it to handle scale up, since it's the only semi-automated platform that is available for all edge-topologies (except SNO atm).

Reviewer 6: It seems like this step is just part of the node joining, which should have happened prior; and this line should be (re)moved. 

That node joining could be UPI, which is the proposal here. Could it potentially be MAO in the future? I know we are trying to keep the scope to UPI joining, but if possible I would like to confirm the design does not prevent future MAO use.

Reviewer 2: Fine for the first pass, but being able to deal with load balancer needs to be adressed soon. With platform:baremetal or any cloud provider, the load balancer needs to be handle the same way as on a day1 installation. Esp. on baremetal, the internal loadbalancer using keepalived etc. has to be deployed. That meenas OTTO needs to ask / be configured with VIPs for ingress and API. This is key, as at the edge, there usually is no external LB available. 
This design and concept needs to consider and prepare for this.

Author: For the initial implementation on `platform: none`, the transition controller does not interact with platform infrastructure. The administrator handles external networking. Platform-specific transitions (baremetal, cloud) are future work.

5. The CLI creates a transition CR with the provided configuration
6. OTTO validates preconditions:

>    - Current topology is SingleReplica

Reviewer 2: Which topology are we talking about here? ControlPlane or Infra? I think this should be made explicit, as those are two different type of changes. WIth SNO>Compact tranistion, we need to handle BOTH.

Author: Resolved by renaming the field to `desiredControlPlaneTopology`. The API section now includes a 'Mapping to status fields' paragraph explaining that for SNO → HA compact, both `controlPlaneTopology` and `infrastructureTopology` transition to `HighlyAvailable`.

   - Target topology is HighlyAvailable
   - Required nodes are reachable and meet minimum resource requirements
   - Required certificates and secrets are present
   - Platform is supported (`platform: none` in the initial implementation)

##### During Transition

7. OTTO signals that a transition is in progress (status on the transition CR)
8. OTTO triggers setup changes on dependent operators:

>    - cluster-etcd-operator (CEO) scales etcd members sequentially following the bootstrap pattern (1→2→3)

Reviewer 3: Does etcd operator need to be told to do this today? How do you tell it to scale up?

Reviewer 4: This is trying to capture the context from a design discussion we had on the previous EP. Essentially, there was a discussion about whether we needed the 1->3 transition to be atomic or whether we should proceed with sequential, single-learner promotion for etcd to scaleup (which is done automatically during bootstrap). The outcome of that discussion was that we'd use the latter, because the atomicity would be a multi-year development effort in the upstream with a low probability of success, and gains you minimal protection over what is already available using quorum-restore.sh today.

To answer your question, CEO handles this automatically, no changes here.

Author: CEO handles etcd scaling automatically via its existing unsafe scaling path on SNO. No explicit signal from the transition controller is needed — CEO detects new control-plane nodes and adds etcd members. The transition controller waits for a healthy 3-member cluster.

   - Each new node joins as a learner and is promoted to a voting member before the next is added
   - The kube-apiserver, kube-controller-manager, and kube-scheduler start on new nodes via static pods managed by the kubelet

>    - Ingress, networking, and other infrastructure operators prepare for multi-node operation

Reviewer 3: Is "prepare" detailed somewhere? Wondering what preparation actually means here?

Reviewer 4: This is bad phrasing. Either the operators pass the pre-flight checks or they don't.
Here's a real example with ingress.
- If you scale up from 1 to 2 control plane nodes, but have single-replica set for ingress - when should it scale up to 2 replicas?
- Pre-flight checks just verify that 3 nodes are available. Ingress continues to operate with 1 replica.
- Pre-flight checks are accepted and OTTO updates the infrastructure config. The control-plane topology is now set to HA.
- Ingress reconciles that change to infrastructure topology and scales up to 2 replicas.
- OTTO runs post-flight checks to ensure the cluster has transitioned successfully (all operators healthy)
- OTTO has nothing to reconcile, so the manifest for the controller is scaled down to 0.

Author: Removed the vague 'prepare' language. The workflow now describes concrete steps — operators react to node presence (step 8), then to topology status changes (step 10).

9. OTTO validates that all operator-specific setup steps have completed successfully
10. OTTO updates the Infrastructure status fields:
    - `controlPlaneTopology` transitions from `SingleReplica` to `HighlyAvailable`
    - `infrastructureTopology` transitions from `SingleReplica` to `HighlyAvailable`

> 11. Operators reconcile against the new topology values and adjust their deployment strategies, replica counts, and placement policies

Reviewer 3: This step is what triggers step 8 no? Why is it after step 8?

Reviewer 4: Step 8 needs to be removed. It makes it seem as thought operators will pre-transition before the transition is admitted.  That's not how this works. We check dependencies for the transition, admit the transition, and verify the cluster reconciles (and roll-back if it doesn't)

Author: Fixed in prior revision — steps are now correctly ordered.

> 11. Operators reconcile against the new topology values and adjust their deployment strategies, replica counts, and placement policies

Reviewer 5: Are we going to try to e.g. restart OLM operators (which previously have treated the topology as fixed)?

Reviewer 3: Do you have a view into how many/which olm operators are reading this value? Are they reading it at startup, or watching the resource? The expected pattern would be that the operator sees the change, and then reacts by updating the operand (e.g. scaling from 1 to 2 replicas now that it's been told the cluster is HA)

Author: OLM operator behavior is listed as an open question. Operators that watch the infrastructure CR should react automatically. Those that read at startup may need a restart. The scope of affected operators needs investigation.

Author: @eggfoobar has been working on some manual demos for this functionality. He might have some insight into the operators based on his runs

Author: The primary operator we discovered that needed to be restarted/updated is ingress. Currently it gets the topology at startup and doesn't handle the reconciliation when new nodes are added.

##### Post-Transition

12. OTTO validates that all operators have reconciled to a healthy state
13. The transition CR is updated to reflect completion
14. The CLI reports success to the administrator

##### Failure Handling

If a transition fails partway through:

- OTTO reports the failure state on the transition CR with diagnostic information

> - For etcd scaling failures, CEO attempts to roll back to the previous member count (e.g., roll back to 1 member if the 1→2→3 scale-up fails)

Reviewer 3: Does this already exist?

Reviewer 4: Etcd is one of the few operators that already scales with the number of control plane nodes regardless of what topology is set.

The key thing that we wanted to highlight here is - let's say the cluster loses quorum during the transition - we restore the API with quorum-restore.sh and restore the previous state.

Reviewer 3: > we restore the API with quorum-restore.sh and restore the previous state

Who does this? Running quorum-restore automatically feels well beyond what I would feel safe implementing as part of an openshift operator. The risk here is incredibly high and if we got to the point where this needs to be done, I'd want the cluster admin to be actioning it

Author: Ack, might be worth noting that recovery steps are outside the scope of this operator controller and in the event of known (etcd quorum loss for example) failures, cluster admins should follow documented recovery steps and retry.

Author: Clarified — quorum loss requires manual recovery via `quorum-restore.sh`. The transition controller does not attempt automated recovery. Recovery steps are the administrator's responsibility.

- The administrator can inspect OTTO logs and the transition CR for details
- The administrator can retry the transition after addressing the issue

### API Extensions

#### Transition Custom Resource

OTTO will define a custom resource for managing topology transitions:

```yaml
apiVersion: topology.openshift.io/v1alpha1
kind: TopologyTransition
metadata:
  name: sno-to-ha-compact
spec:
  targetTopology:
    controlPlane: HighlyAvailable

>     infrastructure: HighlyAvailable

Reviewer 3: Why can this not be in the `Infrastructure` spec?

Reviewer 4: It could be. What Jeff and I realized is that the transition operation a relatively niche use case. So our thinking was we could isolate this to its own, fully encapsulated CRD which is your one-stop-shop for transition information. In the end, I think it's a hollow victory because you're going to need to update infrastructure to complete the transition anyway so you might as well consolidate in the existing infra config.

Author: Now using Infrastructure spec — `desiredControlPlaneTopology` is a field in `InfrastructureSpec`.

OTTO updates these fields as the final step of a transition, changing them from one fixed mode to another (e.g., `SingleReplica` → `HighlyAvailable`).

A ValidatingAdmissionPolicy will be added to prevent direct edits to topology fields outside of OTTO's service account. This ensures transitions are always orchestrated rather than applied ad hoc.

#### Feature Gate

A new feature gate `MutableTopology` will be added to gate this functionality. The feature gate will progress through the following stages:

- **Dev Preview**: Part of the `DevPreviewNoUpgrade` feature set
- **Tech Preview**: Moved to the `TechPreviewNoUpgrade` feature set
- **GA**: Moved to the `Default` feature set

### Topology Considerations

#### Hypershift / Hosted Control Planes

Mutable topology is not compatible with HyperShift clusters. HyperShift uses `External` as its `controlPlaneTopology`, and topology transitions are not applicable to hosted control planes where the control plane lifecycle is managed externally.

Future support for HyperShift is not planned for this enhancement but is not ruled out.

> #### Standalone Clusters

Reviewer: What is a standalone cluster?

Reviewer 3: Traditionally this means general HA clusters, provisioned by IPI or UPI

Author: JoelSpeed answered — standalone = general HA clusters provisioned by IPI or UPI.

MicroShift is not affected by this enhancement. A separate enhancement may address MicroShift-to-SNO transitions but it is unlikely it will be included as part of supported transitions in OTTO.

#### OpenShift Kubernetes Engine

This proposal does not depend on features excluded from the OpenShift Kubernetes Engine (OKE) product offering. OTTO modifies core infrastructure components — the Infrastructure API, installer, cluster-etcd-operator, and other in-payload operators — all of which are included in OKE. OKE clusters that use mutable topology will benefit from the same transition capabilities as OCP clusters.

### Implementation Details/Notes/Constraints

#### OpenShift Topology Transition Operator (OTTO)

OTTO is a new optional payload operator with the following characteristics:

- **Ships with the payload** but is **not installed by default**
- Installed either manually or via the `oc adm transition topology` command
- Owns the transition graph — the directed graph defining which topology transitions are supported
- Owns the validation criteria for each transition (required nodes, certificates, secrets, operator states)

> - Orchestrates transitions by interacting with cluster operators via their existing APIs

Reviewer 3: I don't think this actually exists

Reviewer 4: This statement is just objectively incorrect. :( I can see why you'd be confused reading this.
This never happens. What we can do is look to update things like ingress and console to be more adaptable like etcd/api-server such that they update their replicas when more infrastructure nodes become available and do firmer pre-flight checks so that the "transition" piece becomes a no-op, but I think it's OK for some operators to continue to treat the topology field as the source of truth for desired behavior.

An alternative would be to key off the infrastructure topologies "desiredTopology" and update the hooks for ingress to try to update it's replicas when it detects an update to that field. Then the pre-flight checks actually verify that has the right number of replicas and we update the topology after it's already succeeded. i guess it depends on whether we're treating the topology field or the desired topology field as the answer to "what should the operator being doing right now".

Reviewer 3: >  it's OK for some operators to continue to treat the topology field as the source of truth for desired behavior.

Absolutely.

In an ideal world, most operators would not scale up their operands until the status toplogy fields were updated. We know that's not true today but I don't think we necessarily need to fix most of the controllers. The one controller that does concern me is etcd operator. Would be good to understand why it acts the way it does today (will just scale up and add the member to quorum on SNO) and whether there's a way we can change that behaviour so that it would treat new members as learners until the status toplogy transitions

@jeff-roche BTW can we get rid of the objectively incorrect statements at some point please

Author: I meant to remove this, will do so in the next update

Author: Removed in latest push.

Reviewer 3: Doesn't appear to have been removed yet

Author: I swear it's gone now

- Updates the Infrastructure status field as the final step, after the cluster is ready for the new topology
- Reports transition status via custom resources

##### Transition Graph

OTTO maintains a directed graph of supported transitions. For the initial implementation:

```text

> SingleReplica (SNO, platform: none) → HighlyAvailable (3-node compact)

Reviewer 5: I think it's a mistake to define the supported topologies in terms of the `controlPlaneTopology` field. There are at least 6 use cases I can think of that users have articulated:

- single-node (1 schedulable control plane, 0+ workers, no load balancer)
- compact (3 schedulable control plane, 0+ workers)
- standby (3 non-schedulable control plane, 0 workers)
- HA (3 non-schedulable control plane, 2+ workers)
- TNA (2 non-schedulable control plane, 1 arbiter, 2+ workers)
- TNF (2 schedulable control plane w/ STONITH, 0 workers)

Author: I've expanded the detail around CP and infra topology, as well as some validation rules around number of workers. For the first pass, we will report an error prior to transitioning if there are _any_ worker nodes.

Reviewer 3: Can you point me to this expansion? I have the same question as Zane still having re-read the EP. This IMO needs more expansion unless I missed a section

Author: The Terms section now acknowledges the broader topology landscape (TNA, TNF, compact, etc.) and explicitly states this enhancement targets `controlPlaneTopology` transitions only. The architecture doesn't preclude future expansion to other configurations. The field is named `ControlPlaneTopology` to make this scope explicit.

Reviewer 4: In the frontend, I've allowed users to specific control-plane and infrastructure targets separately and believe we should expand in that direction for TP

> ```

Reviewer 6: We shouldn't call this a `DAG`. Technically it is correct at the moment because only a single transition is supported, but future work all points to reverse transition/scale down, in which case we have a cycle, which is fine. We should just call it a directed graph (digraph) or state machine, including "acyclic" is misleading.

Author: Agreed, dropped the acyclic moniker around the file for either "directed graph" or "transition graph" depending on context

Author: Applied in latest push — replaced all DAG/acyclic references with 'directed graph' or 'transition graph'.

#### `oc adm transition topology` CLI Command

The CLI command provides an interactive interface for topology transitions:

- Installs/activates OTTO if not already present
- Guides the administrator through required configuration (nodes, certificates, secrets)
- Creates the transition CR with the provided configuration
- Monitors the transition CR for status updates
- Reports success or failure with diagnostic information

The CLI does not contain transition logic — it delegates entirely to OTTO. This keeps the CLI thin and avoids bloating it as more transitions are supported.

#### etcd Scaling: SNO to HA Compact

When transitioning from SNO to a 3-node compact cluster, CEO scales etcd members sequentially — the same approach used during cluster bootstrapping (1→2→3 members):

1. **Starting state**: 1 etcd voting member (quorum=1)
2. CEO adds an etcd learner on the second control-plane node
3. The learner syncs data from the existing voter via snapshot transfer
4. CEO promotes the learner to a voting member — the cluster now has 2 voting members (quorum=2)
5. CEO adds an etcd learner on the third control-plane node
6. The learner syncs data from an existing voter

> 7. CEO promotes the learner to a voting member — the cluster now has 3 voting members (quorum=2)

Reviewer 2: ```suggestion
7. CEO promotes the learner to a voting member — the cluster now has 3 voting members (quorum=3)
```

Author: If I understand correctly, quorum for a 3-member etcd cluster is 2 (majority = ceil(3/2) = 2), not 3

Reviewer 7: Correct, the 3rd member does not effect quorum, it just allows us to survive losing a member without losing quorum

8. The cluster can now tolerate the loss of one control-plane node

This is a sequential process. The 2-member state in steps 4–5 is the primary risk window — quorum requires both members, so losing either is fatal. This window is minimized by proceeding to step 5 immediately after promotion.

The 2-member state is transient and follows the same pattern as cluster bootstrapping — a well-exercised code path.

#### Component Changes Summary

| Component | Changes Required |
| --------- | ---------------- |
| OTTO (new) | Transition operator with transition graph, validation, orchestration, and transition control CRD(s) |
| `oc` CLI | New `oc adm transition topology` interactive command |
| Infrastructure API | ValidatingAdmissionPolicy to restrict direct topology field edits |
| cluster-etcd-operator | Coordinate with OTTO for sequential etcd scaling during transitions |
| Ingress, networking, monitoring operators | Respond to OTTO coordination signals during transitions; reconcile on Infrastructure config changes |

> #### Platform Support Constraints

Reviewer 5: We need to mention that IBI clusters cannot be converted from SNO (and have some mechanism for preventing that).

Reviewer 3: What's the technical blocker there?

Author: Added IBI exclusion to Topology Considerations and Non-Goals. We haven't evaluated whether there are technical constraints — it's out of scope for this enhancement.

The initial implementation targets `platform: none` clusters. On `platform: none`, the administrator is responsible for managing their own load balancing configuration (VIPs, DNS) when scaling beyond a single node.

> `platform: baremetal` support is planned for a subsequent phase. Bare metal networking uses keepalived for ingress load balancing, which is not useful and creates a point of failure for SNO deployments. The Bare Metal Networking team will be consulted to determine if this networking setup can be enabled for single-node clusters transitioning to HA.

Reviewer 5: I find it weird that we are going to add single-node support to `platform:baremetal` just so that we can say we are not preventing it from later transitioning to HA.
Who is asking for this?

I would prefer that any effort from the on-prem networking team were instead directed toward adding optional on-prem networking to `platform:external`.

Reviewer 2: As said in a previous comment, its crucial to get support for "plaform:baremetal" and keelaived load balancing (for ingress AND API) in the medium term. We should validate that there is no technical obstacle and this can be added in the next release. Rational: at the edge, there hardly is an external load balancer available.

Author: Out of scope for this EP. The Standalone Clusters section now documents the explicit rationale for targeting `platform: none` only. Baremetal is future work pending keepalived resolution.

> ### Risks and Mitigations

Reviewer 5: I would say one is the inability to validate external requirements, such as correct load balancer setup.

Author: Covered by the 'Cannot Validate External Requirements' risk section — external networking is the administrator's responsibility.

#### Risk: Quorum Loss During Two-Member Transient State

**Risk**: During sequential etcd scaling (1→2→3), the cluster passes through a 2-member state where quorum=2. Losing either member during this window causes quorum loss.

**Mitigation**:
- The 2-member state is transient and follows the same sequential pattern used during cluster bootstrapping — a well-exercised code path
- Learner instances are used before promoting members to minimize the promotion window
- No availability guarantee during transitions; administrators should treat scaling operations as a maintenance window

> - CEO will attempt rollback if scaling fails (e.g., rollback to 1 member if the 1→2→3 scale-up fails partway through)

Reviewer 2: What happens if the loss of quorum=2 was created by a split brain situation? Will both etcd attempt rollback to 1? This could lead to two individual clusters of one. I would be fine with specifing a simple heuristic to resolve this situation, e.g. dropping the younger etcd instance in favour of the older one or something like that.  Or maybe a special command for the admin to resolve this situation

Author: Quorum loss during the 2-member window requires manual recovery via `quorum-restore.sh`. Added to failure examples and recovery table. Automated split-brain resolution is not in scope.

> - Future iterations may explore admitting two learners simultaneously and promoting only when both are ready, eliminating the 2-member voting window entirely but that is out of scope for this enhancement

Reviewer 2: maybe worth adressing this directly, instead of dealing with the potential split brain situation from my previous comment?

Author: Covered in the failure examples — network partition between 2 etcd members during scaling is now called out explicitly as a failure scenario requiring manual recovery.

#### Risk: Transition Fails Partway Through

> **Risk**: A transition may fail after some operators have begun reconfiguring but before the transition completes, leaving the cluster in an intermediate state.

Reviewer 3: Can you provide examples of failures?

Reviewer 4: @jeff-roche one specific example:
- etcd scales up to 2 nodes
- network partition between them
- both have lost quorum
- no api

Reviewer 2: I thought operators transition/reconfire only after etcd scaling is complete?
I think a better example would be for e.g. ingress operator to scale up from 1 to 2, but for whatever reasons, ingress pod on the new node can not start, e.g. due to ImagePullBackoff (registry not available)

Author: Added three concrete examples: (1) etcd quorum loss during 2-member window, (2) new node becomes NotReady, (3) operator fails to reconcile post-transition (e.g., ImagePullBackOff).

**Mitigation**:
- OTTO validates preconditions before starting
- OTTO sequences operations so that the Infrastructure config is updated only after all setup steps succeed
- Operators do not see a topology change until OTTO updates the Infrastructure status as the final step
- If setup steps fail, OTTO reports the failure and CEO attempts rollback for etcd
- The transition CR provides detailed status for troubleshooting

#### Risk: Platform Bare Metal May Not Support Single-Node Clusters

> **Risk**: If keepalived networking cannot be enabled, `platform: baremetal` will be limited to 2+ nodes, reducing the value of mutable topology for this platform.

Reviewer 3: limited to 2+? Isn't that the success criteria?

Reviewer 4: baremetal platform doesn't support SNO because having a load balancer for 1 node doesn't make sense.
In order for users to get the benefits of having not having to manually deploy a load balancer (i.e. what they primarily save in terms of effort when deploying on platform: baremetal), we need to investigate if we can allow baremetal as a platform for SNO first (which loadbalancing disabled), and change that operator so that loadbalancing can be introduced post-transition.

Otherwise we need to introduce a new, scarier feature: **platform transitions**.
That a pandora's box I don't want to look at.

Reviewer 3: > That a pandora's box I don't want to look at.

You and me both

So are we tying this EP to not only supporting topology transitions, but also SNO on baremetal? I would have expected a SNO on baremetal project to be sufficiently large and warrant its own EP?

Author: The baremetal risk is now framed as future scope. `platform: none` is the only supported platform for the initial implementation.

> **Risk**: If keepalived networking cannot be enabled, `platform: baremetal` will be limited to 2+ nodes, reducing the value of mutable topology for this platform.

Reviewer 5: This is out of date given the current scope, right?

Author: Updated — the baremetal risk is now framed as future scope with a dependency on the Bare Metal Networking team.

- The `Adaptable` enum value created a paradigm that was fundamentally different from existing fixed topology modes

Mutable topology achieves the same end goal (SNO clusters can grow to HA) with less operator-side complexity. Operators continue to react to the same fixed topology values they already understand. Transition complexity is concentrated in OTTO rather than distributed across all operators.

### CLI-Only Transition Runner

An alternative is to embed all transition logic in the `oc adm transition` command without a dedicated operator.

**Why it was rejected**:
- The set of supported topologies is bounded, so the transition graph itself stays small. However, each transition is a long-running, multi-step process (etcd scaling alone takes minutes) that requires persistent state tracking a CLI process cannot reliably provide — a dropped SSH session or terminal close would leave the cluster in an intermediate state with no automated recovery
- Error recovery and retry logic is better suited to an operator's reconciliation loop than imperative CLI code
- The CLI would need direct access to operator internals, violating separation of concerns

> ### Extending an Existing Core Operator

Reviewer 3: Or cluster config operator which would make a very natural home for this as long as we have commitment of ownership from folks writing the new controller

Reviewer 6: At risk of going on a tangent, currently the installer has problems calculating topology when laying down manifests. We have a bug for this in the backlog, and I left https://github.com/openshift/enhancements/pull/1905#issuecomment-4399210626 on the previous enhancement.

I would like to see that calculation moved to the cluster config operator in bootkube during bootstrapping. That solution could co-exist with this one (and my team will push it forward as priorities allow); but it could potentially also tie into this solution.

Reviewer 4: I'm find with us using CCO for this. I will take the blame for miscommunicating this to Jeff - it didn't strike me as obvious that a controller for this transition would obviously belong there. My instincts were that new code in the core operators is expensive, especially for a controller that doesn't need to be running 99% of the time. That said, I think it's fine for this to be a controller that is installed with zero replicas and the replicas are scaled-up during transition events. That fits them main sentiment of what Jeff and I were trying to solve - minimizing the tax on clusters that will never use this feature (i.e. the vast majority of them).

Reviewer 5: IMHO the solution for the installer is to get the user to specify their intent in the install-config (this should be passed through to the cluster). This enhancement is a good opportunity to define what that input should look like.

Author: Resolving this thread as I've re-scoped this to be a new CCO controller

Author: Resolved — controller now lives in CCO. This was the outcome of the discussion with jaypoulz.

Standard downgrade procedures apply. OTTO and the transition CRD are preserved. Completed transitions are not reverted — the cluster retains its current topology.

**Y-stream downgrades** (to a minor version without mutable topology support):
The CVO will evaluate the feature gate during downgrade. If the target release does not include the `MutableTopology` feature gate:
- OTTO will not be managed by the target release's CVO
- Completed transitions are not affected — the Infrastructure config contains standard topology values that the target release understands
- In-progress transitions should be completed or rolled back before downgrading

## Version Skew Strategy

Mutable topology is gated by the `MutableTopology` feature gate. OTTO is only active when the feature gate is enabled.

Version skew during transitions is not a concern because OTTO orchestrates the entire sequence within a single cluster version. Administrators should not initiate upgrades while a transition is in progress.

Post-transition clusters use standard topology values that all operator versions understand. There is no version skew risk for completed transitions.

## Operational Aspects of API Extensions

> OTTO introduces a `TopologyTransition` CRD. This CRD:

Reviewer 5: How about just a Topology CRD? We should store the intent of the user, not just when the cluster is in transition.

Reviewer 3: Infrastructure spec should hold this rather than a separate CRD - since we already have topology as infrastructure status today

Author: Agreed with JoelSpeed — using Infrastructure spec rather than a separate CRD, since topology is already represented in infrastructure status.

**Compact Cluster** — A cluster where control-plane nodes also serve as workers. In the initial SNO-to-HA transition, the target is a 3-node compact cluster with no dedicated worker nodes. The compact deployment shape is a consequence of not adding dedicated worker nodes — it is not a distinct `TopologyMode` enum value.

**mastersSchedulable** — A field in the infrastructure status indicating whether control-plane nodes are schedulable for general workloads. The topology transition controller recalculates this value as part of a transition.

**Cluster Administrator** — An entity responsible for managing an existing cluster, including Day 2 operations such as topology transitions and node scaling.

## Summary

This enhancement enables OpenShift clusters to transition between topology modes as a Day 2 operation. This changes the existing OpenShift assumption that topologies are immutable after installation.

> A new `desiredTopology` field in the infrastructure spec expresses the administrator's intent to transition. A topology transition controller in cluster-config-operator watches for changes to this field, validates preconditions, coordinates the transition, and updates the existing topology status fields when the cluster is ready.

Reviewer 3: Is this for infrastructure or control plane, or both?

Author: This is a fair question. In my head this entire process is about control plane scaling. I think we already have the necessary mechanisms in place to scale workers, right?

Author: Resolved by renaming to `desiredControlPlaneTopology` — makes the scope explicit. The controller derives `infrastructureTopology` and `mastersSchedulable` from the transition definition.

> A new `oc adm transition topology` CLI command provides an interface for cluster administrators to initiate transitions.

Reviewer 3: Is this a common addition to the CLI? I have nothing against extending the CLI, but do question if it is strictly required

Author: I think it is not strictly required, this is more of a usability thing. In theory a cluster admin could go in and update the desired topology and manually monitor progress but that might feel disconnected. Through the CLI we could give some structure to the process

Author: The CLI is a convenience, not strictly required. An administrator could patch `spec.desiredControlPlaneTopology` directly. The CLI follows the `oc adm upgrade` pattern — validates preconditions client-side and patches the CR.

Reviewer 4: The CLI is about ensuring a seemless UX. It also gives us a chance to pre-warn the user for things that may cause issues before we even admit the change. For example, if we don't have the correct number of control-plane nodes and etcd members, the CLI warns the user and forces a separate acknowledgement flag to ignore the warnings.

This works similar to upgrades (although upgrades do all validation server side, which is something we should move towards a s well.)

We follow a dry-run + confirm pattern. Other oc commands that follow this pattern are:
- oc adm prune builds
- oc adm migrate storage
- oc adm groups sync

Reference PR is here:
https://github.com/openshift/oc/pull/2295

> The initial implementation supports transitioning Single Node OpenShift (SNO) clusters to HA compact (3-node) on `platform: none`.

Reviewer 3: Hoping to see somewhere a documented reason for why we are only considering platform none

Author: ack, I think this is covered a couple times in this doc but I can find a more explicit place to mention the reasoning

Author: Added explicit rationale in the Standalone Clusters section — edge computing is the primary customer, `platform: none` has no platform-managed infrastructure to coordinate, baremetal needs keepalived resolution, cloud platforms need CCM integration.

* Officially support topology transitions in OpenShift
* Provide a supported interface for administrators to initiate topology transitions
* Support transitioning SNO clusters to HA compact (3-node) on `platform: none` as the initial transition path
* Maintain backward compatibility — existing clusters with fixed topology modes are unaffected
* Establish the architectural foundation for additional transition paths in the future

### Non-Goals

* Supporting all possible topology transitions in the initial implementation (only SNO → HA compact on `platform: none`)
* Supporting transitions for HyperShift or hosted control plane clusters
* Supporting transitions for MicroShift deployments
* Automatic node provisioning or deprovisioning based on workload demands
* Scaling down control-plane or worker nodes (scale-down may be addressed in a future enhancement)
* Supporting bidirectional transitions (e.g., HA → SNO) in the initial implementation

## Proposal

This enhancement introduces a new infrastructure API field and a topology transition controller in cluster-config-operator (CCO; not to be confused with cloud-credential-operator) to enable topology transitions as Day 2 operations.

> The approach follows the standard Kubernetes spec/status contract and mirrors the pattern used by `oc adm upgrade`:

Reviewer 3: It's more an openshift thing rather than a kube thing this pattern

Author: ack, will update wording

Author: Fixed — changed to 'OpenShift'.

1. **`desiredTopology` field in InfrastructureSpec** — Expresses the administrator's intent to transition. The CLI patches this field to initiate a transition. The existing `controlPlaneTopology` and `infrastructureTopology` fields in status continue to represent the cluster's observed topology.

2. **Topology transition controller in cluster-config-operator** — A new controller in CCO that watches the infrastructure CR for `desiredTopology` changes, validates preconditions, coordinates the transition, and updates the status topology fields when the cluster is ready for the new mode.

3. **`oc adm transition topology` CLI command** — A command that validates preconditions before patching `spec.desiredTopology` on the infrastructure CR, then monitors transition progress.

> The transition controller is proposed to live in cluster-config-operator because CCO is the canonical location for config.openshift.io CRD manifests and bootstrap CR rendering, and the topology transition logic is tightly coupled to the Infrastructure CR schema it ships. This is a deliberate expansion of CCO's scope since historically the repo has been limited to CRD manifests and bootstrap rendering. The controller is feature-gated using the standard library-go FeatureGateAccess pattern: when the gate is disabled the controller is not registered with the manager and incurs negligible runtime overhead; a gate change triggers an operator restart via ForceExit so the new state is picked up cleanly.

Reviewer 3: > the repo has been limited to CRD manifests and bootstrap rendering

This is not really true, but also doesn't materially affect what you're trying to say in this EP

TBH, this whole paragraph is fluff IMO

Author: Good with me dropping it? It was a recommendation from chai-bot to add it and I figured it didn't hurt but agree it's fluff

Author: Agreed — removed the CCO scope history paragraph. Kept the canonical-owner justification.

#### OpenShift Kubernetes Engine

This proposal does not depend on features excluded from the OpenShift Kubernetes Engine (OKE) product offering. Mutable topology modifies core infrastructure components — the infrastructure API, cluster-config-operator, cluster-etcd-operator, and other in-payload operators — all of which are included in OKE.

### Workflow Description

#### Transition: SNO to HA Compact (3-Node)

**Operational guidance**: Administrators should treat topology transitions as a maintenance window. The cluster is expected to remain available throughout the transition, but availability is not guaranteed — administrators should reduce non-critical workload risk accordingly. The cluster is expected to be fully available before and after the transition.

##### Pre-Transition

1. The cluster administrator prepares at least 2 additional control-plane nodes and joins them to the cluster — the kubelet is running on each node and Node objects exist in the Kubernetes API. On `platform: none`, the administrator manages their own load balancing configuration (VIPs, DNS).
2. The cluster administrator runs `oc adm transition topology HighlyAvailable`

> 3. The CLI validates preconditions before patching (e.g., feature gate enabled, no transition already in progress)

Reviewer 3: Why does a transition in progress matter? If I initiated a transition and then decided I wanted to cancel it, I would expect there is a point up to which I could do this (before the status is updated?)

Author: I'll drop that wording, I think any transition not already being in progress (I think doing an upgrade would count here) is already going to be blocked by the API, the CLI can be dumb here

Author: Clarified — cancel is possible before the status fields are updated (step 9) by resetting `desiredControlPlaneTopology` to match current topology. After the status update, the transition is effectively complete and cannot be cancelled.

Reviewer 3: This wording doesn't look like it was updated to reflect that

Author: @JoelSpeed are you referring to the CLI command specifically?

Reviewer 4: The CLI status command I added is a convenience. It monitors the state of the CCO conditions and gives an easy way to see the relevant conditions at a glance. It's more like syntactic sugar, but I do think it improves the UX. While you can't technically undo, you could in theory start multiple transitions in the future. (CP -> HighlyAvailableArbiter, Infra -> HighlyAvailable). IDK if we want to block multi-inflight transitions from a CCO perspective, but my observation is that it's helpful to be able to see what is in the process of changing without having to know exactly what to filter out of the CCO conditions.

4. The CLI patches the infrastructure CR: `spec.desiredTopology: HighlyAvailable`
5. The API server validates `desiredTopology` via CEL validation rules, rejecting unsupported topology modes before accepting the write

##### During Transition

6. The topology transition controller in CCO detects the `desiredTopology` change and validates preconditions:
   - Current `controlPlaneTopology` is `SingleReplica`
   - Target topology is `HighlyAvailable`
   - At least 3 nodes with `node-role.kubernetes.io/control-plane` or `node-role.kubernetes.io/master` labels are present in the Node API
   - Platform is supported (`platform: none` in the initial implementation)

> 7. The controller signals that a transition is in progress (via infrastructure status conditions)

Reviewer 3: No? It signals the transition is allowed by updating the status fields (step 9) and then its job is done. The other operators (that haven't already reacted) now react and update their config

Author: I think my wording here could be better. What it's actually doing here is updating a status field in CCO to say that a tranistion is in progress (or being prepared)

Author: Reframed in latest push. The orchestration section now describes 'Wait for etcd scaling' and 'Wait for operator reconciliation' rather than active coordination. Added a note that future transitions (e.g., TNF) may need explicit pre-update coordination steps.

8. Operator-specific setup:

>    - cluster-etcd-operator (CEO) scales etcd members sequentially (1→2→3), reusing the learner-to-voter promotion mechanism from bootstrapping — each new node joins as a learner and is promoted to a voting member before the next is added

Reviewer 3: Have you manually verified that this works? Does anything actually need to be changed right now?

AFAICT when I looked through the code last week, while the cluster is in the SingleReplica toplogy, the etcd operator uses its unsafe scaling method. Which means that as soon as I add the new control plane machine, it will scale a new etcd pod out, and add it to the cluster as a voting member.

The only thing I could see that was different based on topology was the deployment of the etcd-guard PDB.

This needs to be verified

Author: I know @jaypoulz did some manual scale testing at the onset of the TNF project and we've heard from @dusk125 and @tjungblu that this should can be done. I know Jeremy has it on his plate to do some additional validation in an upcoming sprint (after TNF GAs so he has more time)

Author: This needs verification. Jeremy has it on his plate to do additional validation. CEO's unsafe scaling path on SNO clusters is a critical dependency for this proposal — the exact behavior needs to be confirmed before entering dev preview.

Author: We've confirmed this works

>    - The kube-apiserver, kube-controller-manager, and kube-scheduler operators create static pod manifests for the new control-plane nodes; the kubelet runs the pods

Reviewer 3: Technically this happens as part of step 1 of the pre-transition

Author: I will move this to be part of that

Author: Fixed — node joining is in pre-transition step 1.

>    - Ingress controller adjusts replica count based on the new topology (specific scaling behavior to be validated during dev preview)

Reviewer 3: Any other controllers you want to call out? I'm sure there are others that have HA operands that will react to this topology change no?

Author: There definitely will be, these were just the ones we thought would be good to call out. @jaypoulz @eggfoobar did I miss any we've discussed?

Author: The full set of operators with topology-dependent behavior has not been enumerated. Building the per-operator topology dependency matrix is now a prerequisite for entering dev preview — see Graduation Criteria.

> 9. The controller updates the infrastructure status fields:

Reviewer 3: This is actually step 7

Author: This is meant to exist here, The operator notes above are meant for doing any needed pre-work before we can let the cluster auto-reconcile. The best example comes back to TNF with any pacemaker config that might need to be setup. Once any pre-work is completed, CCO will then go and update the infra topology fields with the new value and let the cluster operators reconcile the change

Author: The step numbering has been revised. Step 8 now describes node-driven operator reactions (etcd, static pods) and step 10 describes topology-driven operator reactions after the status update. The pre-transition configuration step is preserved for future transitions (TNF).

   - `controlPlaneTopology` transitions from `SingleReplica` to `HighlyAvailable`
   - `infrastructureTopology` transitions from `SingleReplica` to `HighlyAvailable` (no dedicated workers, so it matches control plane topology)

>    - `mastersSchedulable` is recalculated

Reviewer 3: What does this actually mean?

Author: I could use some help defining this. This is kind of a placeholder for what @patrickdillon mentioned previously

Author: Removed the vague placeholder. Steps now describe concrete actions.

> 10. Operators reconcile against the new topology values and adjust their deployment strategies, replica counts, and placement policies

Reviewer 3: This step is the same as step 8 no?

Author: Nope, step 8 is deliberate config that we may need to do (what that entirely looks like is still being discovered). This is the auto-reconcile period after the infra was updated

Author: Deduplicated. Steps 8 and 10 are now clearly distinguished — step 8 covers node-driven reactions (before status update), step 10 covers topology-driven reactions (after status update).

>     **Note**: OLM-managed operators that read topology at startup rather than watching for changes may need to be restarted after the transition completes. See [Drawbacks](#drawbacks) for details.

Reviewer 3: Is there a standard library function that OLM based operators leverage to watch this value and understand how they should behave? If there is, it would be preferential to adjust it to watch the value and restart the operator when it detects a change (much like we do for feature gates)

Author: I'd anticipate OLM operators would watch for changes to the infrastructure status fields, I think this is already possible in library-go if I'm not mistaken

Author: This needs investigation. The OLM operator impact is listed as an open question — we need to determine which operators watch the infrastructure CR vs read at startup.

Reviewer 3: @joelanford do you have any idea what add-ons might be doing here?

Author: Operators that need to react to the topology changing can already watch it (the same way we watch the spec field in the CCO controller to make changes) and react to the change. As is, the majority (excluding ingress) of cluster operators already gracefully handle the transition.

##### Post-Transition

> 11. The controller validates that critical operators have reconciled to a healthy state

Reviewer 3: And does what with this information?

Author: If they've successfully reconciled, nothing. On error it will report the error and attempt a rollback

Author: So I guess also update the new CCO status field reflecting current transition state

Author: If operators have successfully reconciled, the controller sets `TopologyTransitionCompleted`. On error it sets `TopologyTransitionFailed` with diagnostic information. The conditions are on the CCO ClusterOperator status.

Reviewer 3: Failed implies a terminal failure, these are very rare in Kube. In what case could the cluster reach a terminal failure in topology transition?

Maybe we want to consider something for a condition that describes the topology without using Complete or Failed terminal states?

> 12. The infrastructure status reflects the completed transition. `desiredTopology` matches `status.controlPlaneTopology`, so no further action is taken.

Reviewer 3: This has already happened

Author: Yeah, this step is more of a note that would likely exist in documentation for cluster admins. I'm not sure the best way to represent that the two matching is what signals the transition has happened

Author: Fixed — step 12 now describes the completion condition, not a redundant action.

> 13. The CLI reports success to the administrator

Reviewer 3: Oh the CLI is expected to wait for all this to happen? What is the CLI looking for to report success?

Author: It's going to monitor a new status field in CCO, I think I should make it more clear that in addition to a new controller in CCO, there would be an accompanying status field

Author: Fixed — CLI now returns immediately after patching. Administrators monitor progress separately via `oc get clusteroperator cluster-config-operator -o yaml` or a dedicated status subcommand (exact UX TBD during dev preview).

##### Failure Handling

> If a transition fails partway through:

Reviewer 3: Please expand on how it's possible to fail, so far the only thing I can think of is etcd losing quorum somehow (and even that, how and when is that possible)

Author: ack, will add more detail around this. Currently all I can think of is etcd but I also don't know all the core operators intimately (yet)

Author: Expanded — added concrete failure examples (quorum loss, node readiness, operator reconciliation). Also clarified throughout that quorum loss = manual recovery, not automated rollback.

> - The controller reports the failure via infrastructure status conditions with diagnostic information

Reviewer 3: Please define somewhere the conditions you expect this controller to now be setting

Author: Defined — the conditions table in the API section lists `TopologyTransitionProgressing`, `TopologyTransitionCompleted`, and `TopologyTransitionFailed` on the CCO ClusterOperator status. The orchestration steps reference when each is set.

- For etcd scaling failures, CEO attempts to roll back to the previous member count (e.g., roll back to 1 member if the 1→2→3 scale-up fails)
- The administrator can inspect CCO logs and infrastructure status for details

> - On failure, the controller resets `desiredTopology` to the pre-transition topology value so that `desiredTopology` matches `status.controlPlaneTopology` again (idle state). This is a deliberate deviation from the standard Kubernetes convention that controllers should not mutate spec: without the reset, the controller would continuously retry a failed transition with no administrator intervention, which is undesirable for a high-risk orchestration workflow. By resetting spec, the controller returns to an idle state and the administrator can address the underlying issue and explicitly re-initiate the transition via `oc adm transition topology`

Reviewer 3: You shouldn't do this. That field is set by the admin, whether we reached that valid state or not is not our concern. The controllers should continually try to reach that state until the user changes the desired value.

In theory though this should never happen.

Author: ack, I was back and forth on the pattern to use for this knowing just reporting the error and letting the admin reset it is the more normal pattern. I'll switch over to that for this paragraph.

Author: Switched to the standard K8s pattern — `desiredControlPlaneTopology` remains unchanged on failure, controller retries with exponential backoff. Administrator cancels by resetting the field. Also clarified that cancel is only possible before the status update (step 9); after that, the transition is effectively complete.

### API Extensions

#### Infrastructure API Changes

This enhancement modifies the existing infrastructure CR (`infrastructures.config.openshift.io`) following the standard Kubernetes spec/status contract:

**Spec (user intent):**

> A new `desiredTopology` field is added to `InfrastructureSpec` to express the administrator's intent to transition:

Reviewer 3: I expected the topology in the spec to mirror the status, so users would set the control plane and infrastructure topology separately. How do you translate a single value in the spec to the intent of three fields (including masters schedulable) in the status?

Author: The current implementation was only going to reflect desiredTopology of the control plane. I think we already have mechanisms for scaling workers, no?

Author: Renamed to `desiredControlPlaneTopology` to make the scope explicit. Added a 'Mapping to status fields' paragraph explaining how the controller derives `infrastructureTopology` and `mastersSchedulable` from the transition definition.

```go
type InfrastructureSpec struct {
	CloudConfig  ConfigMapFileReference `json:"cloudConfig"`
	PlatformSpec PlatformSpec           `json:"platformSpec,omitempty"`
	// desiredTopology expresses the administrator's intent for the
	// cluster's control plane topology. The installer sets this field
	// to match the cluster's initial controlPlaneTopology. When the
	// value differs from status.controlPlaneTopology, the topology
	// transition controller in cluster-config-operator initiates a
	// transition. When the value matches status.controlPlaneTopology,
	// no transition is in progress.
	// +openshift:enable:FeatureGate=MutableTopology
	DesiredTopology TopologyMode `json:"desiredTopology"`
}
```

> There is no kubebuilder default — the initial value is not static. The installer (or a CCO bootstrap reconciler) writes `desiredTopology` to match the cluster's installed `controlPlaneTopology` (e.g., `SingleReplica` for SNO, `HighlyAvailable` for standard clusters). This makes the field effectively required-on-create at install time. A transition is initiated when `desiredTopology` differs from `status.controlPlaneTopology`. No transition occurs when the values match.

Reviewer 3: IMO it should just be empty _until_ an admin has set a value. That's useful because it allows us to see whether an admin has ever attempted to conduct a transition. If the value is always empty (as it would be on an upgraded cluster), we know no one has tried a transition

Author: ack, will swap to that. I like that better.

Author: Done — field is now `+optional`, `omitempty`, empty by default. Installer does not populate it. Empty = no transition ever requested.

**Status (observed state):**

The existing fields in `InfrastructureStatus` that the controller updates upon successful transition:

```go
// controlPlaneTopology expresses the expectations for operands that normally
// run on control nodes. Currently documented as "set once by the installer
// and not expected to change." This enhancement changes that contract when
// the MutableTopology feature gate is enabled.
// +kubebuilder:default=HighlyAvailable
ControlPlaneTopology TopologyMode `json:"controlPlaneTopology"`

// infrastructureTopology expresses the expectations for infrastructure
// services that do not run on control plane nodes. When there are no
// dedicated worker nodes, this is set to match controlPlaneTopology.
// +kubebuilder:default=HighlyAvailable
InfrastructureTopology TopologyMode `json:"infrastructureTopology,omitempty"`
```

No new enum values are added to `TopologyMode`. The existing values (`SingleReplica`, `HighlyAvailable`, `DualReplica`, `HighlyAvailableArbiter`) are sufficient.

**Transition progress** will be reported via the following condition types on the infrastructure status:

| Condition Type | Meaning |
| -------------- | ------- |

> | `TopologyTransitionProgressing` | A transition is in progress. `status: True` when actively transitioning, `status: False` when idle or complete. |

Reviewer 3: Can you define when this gets set to true and when it gets set to false in more detail?

For example, does it get set to true as soon as the spec differs from status? Or only once the status fields have been updated? There is actually no transition happening until the status fields are updated, so IMO it would be correct to change then

But if you do that, when does it get set back to false? Is CCO supposed to find every component in the cluster and understand if it is transitioning?

Author: I will update the steps above to reflect the status field updates

My thought is we would check the list of core operators to make sure they successfully transitioned. I haven't fully thought out the how on that yet.

Author: Documented in the orchestration steps — `TopologyTransitionProgressing` is set at step 2 (signal transition), cleared at step 6 (report completion). The conditions table in the API section defines the full lifecycle.

| `TopologyTransitionCompleted` | The most recent transition completed successfully. `status: True` after successful completion. |
| `TopologyTransitionFailed` | The most recent transition attempt failed. `status: True` when a failure has occurred. `message` contains diagnostic details. |

These condition types provide a stable contract for the CLI, console, and telemetry consumers. Reason values (e.g., `TransitionStarted`, `EtcdScalingInProgress`, `WaitingForOperators`, `PreconditionNotMet`, `EtcdScalingFailed`) will be refined during dev preview implementation.

#### Admission Control

> **Spec validation**: A CEL validation rule on the `desiredTopology` field will restrict accepted values to topology modes that have defined transitions. For the initial implementation, this limits the field to `SingleReplica` and `HighlyAvailable`. Setting `desiredTopology` to an unsupported value (e.g., `External`) will be rejected by the API server. Access to `spec.desiredTopology` is governed by the existing RBAC for the infrastructure CR (`infrastructures.config.openshift.io`). By default, only users with `cluster-admin` or equivalent roles can modify infrastructure spec fields. No additional RBAC restrictions are proposed for the initial implementation; a dedicated role for topology transitions may be considered in future iterations if finer-grained access control is needed.

Reviewer 3: You don't need CEL to restrict it to two values, use a new type alias that only has two possible enum value choices

Author: Done — replaced CEL with a `DesiredTopologyMode` named type using `+kubebuilder:validation:Enum=SingleReplica;HighlyAvailable`. API server rejects invalid values at admission.

> **Status protection**: A ValidatingAdmissionPolicy (VAP) targeting the `/status` subresource of the infrastructure CR will prevent direct edits to the `controlPlaneTopology`, `infrastructureTopology`, and `mastersSchedulable` status fields outside of the CCO topology transition controller's service account (in the `openshift-config-operator` namespace). This ensures transitions are always orchestrated rather than applied ad hoc.

Reviewer 3: I'm not convinced we need to be this defensive. We aren't today for any other fields, we also don't actually gate against most fields in the infrastructure status being immutable. If someone deliberately goes and changes these values and breaks their cluster, that's on them

Author: ack, if we are all good with keeping that as is, I'm good with it. I just figured this was a good opportunity to change that if we wanted to.

Author: Agreed — dropped the VAP entirely. Status fields are unprotected, consistent with other infrastructure status fields.

> The VAP will use a **fail-closed** policy — if the VAP is unavailable, all edits to the protected status fields are blocked. This prevents uncontrolled topology changes when the admission infrastructure is degraded. Note that fail-closed assumes the API server is healthy enough to evaluate admission policies; if the API server itself is unavailable, no writes to the infrastructure CR can occur regardless.

Reviewer 3: What does it mean for a VAP to be unavailable?

Author: Moot — VAP has been dropped entirely.

> ValidatingAdmissionPolicy GA'd in Kubernetes 1.30 (OCP 4.17). VAP support for `/status` subresource matching via `matchPolicy` configuration on `ValidatingAdmissionPolicyBinding` will be validated against the target OCP release during dev preview. During cluster bootstrap, the infrastructure CR status fields are written by the installer before the admission infrastructure is fully initialized — the fail-closed policy does not apply at that stage because the API server's admission chain is not yet active. Post-bootstrap, the VAP exempts the CCO topology transition controller's service account via a `matchConditions` CEL expression.

Reviewer 3: > VAP support for `/status` subresource matching

Yes it does support status subresources

> before the admission infrastructure is fully initialized

This is not true

> the fail-closed policy does not apply at that stage because the API server's admission chain is not yet active.

Also not true

The ordering of installation is what might come into play here, but nothing to do with whether the API servers admission infrastructure is ready or not

Author: Moot — the entire VAP section has been removed. Thank you for the corrections on status subresource matching and admission chain initialization.

- Reports transition progress via infrastructure status conditions

##### Supported Transitions

For the initial implementation:

```text
SingleReplica (SNO, platform: none) → HighlyAvailable (3-node compact)
```

Future transitions can be added without modifying the core controller logic. Each supported transition defines:

- **Preconditions**: What must be true before the transition can start
- **Orchestration steps**: What the controller coordinates during the transition
- **Validation criteria**: What must be true after the transition for it to be considered complete

##### Transition Orchestration

When `spec.desiredTopology` differs from `status.controlPlaneTopology`, the controller follows this sequence:

1. **Validate preconditions** — check that the source topology, target topology, platform, and control-plane node count (3+) are valid
2. **Signal transition in progress** — set infrastructure status conditions and set `Upgradeable=False` on the CCO `ClusterOperator` with reason `TopologyTransitionInProgress` to prevent CVO from initiating an upgrade while the cluster is in an intermediate topology state

> 3. **Coordinate operator-specific setup**:

Reviewer 3: This is in the wrong place. Step 4 comes before this, and the controller doesn't actually do anything for co-ordination, the other operators just react

Author: In the case of SNO -> HA Compact I think this is true. Future iterations I think this is not true. TNF as has been mentioned a few times is going to rely on this step for configuration in future iterations of this. I think it would be remiss to not mention the need for a pre-transition configuration step even if that step doesn't end up being needed for SNO to HA Compact

Author: Reordered in latest push: etcd wait → status update → operator reconciliation wait. The coordination framing is gone for the initial SNO → HA transition. Added a note that future transitions (TNF) may need explicit pre-update coordination.

   - CEO scales etcd members sequentially (1→2→3)
   - Wait for kube-apiserver, kube-controller-manager, and kube-scheduler to start on new nodes
   - Specific requirements for ingress, networking, monitoring, and other operators will be validated and documented during dev preview implementation. These operators are expected to reconcile on infrastructure status topology field changes, but the exact behavior for each has not yet been confirmed.
4. **Update infrastructure status** — change `controlPlaneTopology`, `infrastructureTopology`, and `mastersSchedulable` to the target values

> 5. **Validate post-transition health** — confirm operators reconcile successfully against the new topology

Reviewer 3: How will you monitor the other operators? Do you have an exhaustive list, and what are you looking for to say whether they have transitioned, and whether they have transitioned successfully

Author: Fair question, that is definitely TBD and we need an answer to this.

Author: The monitoring approach is TBD — it will be defined as part of the per-operator topology dependency matrix work, which is a dev preview prerequisite. The orchestration step 5 describes the intent (controller monitors operator health) but the specific signals per operator need validation.

6. **Report completion** — update infrastructure status conditions

#### `oc adm transition topology` CLI Command

The CLI command provides an interface for topology transitions:

- Validates preconditions client-side (feature gate enabled, no transition in progress)
- Patches `spec.desiredTopology` on the infrastructure CR
- Monitors infrastructure status conditions for progress

> - Reports success or failure with diagnostic information

Reviewer 3: How long is a transition expected to take? Are you sure you want this to sit and wait by default? Wondering if users might assume that the CLI command hanging and waiting means the transition is stuck and/or that if they cancel the CLI command it might cancel the transition?

Making the CLI something that returns immediately signals to the user that the transition has been requested, maybe there's a separate subcommand for waiting they can run separately?

Author: A separate sub command might be a good idea. We could also direct them to use oc get with a watch on the new status fields for CCO. That will be the source of truth on where we are in the transition

Author: Agreed — CLI now returns immediately after patching. Monitoring is separate, either via `oc get clusteroperator cluster-config-operator` or a dedicated status subcommand. Exact UX TBD during dev preview.

The CLI does not contain transition logic — it delegates entirely to the CCO controller. This follows the same pattern as `oc adm upgrade`, which patches `spec.desiredUpdate` and lets the CVO do the work.

#### etcd Scaling: SNO to HA Compact

When transitioning from SNO to a 3-node compact cluster, CEO scales etcd members sequentially. Each new member joins as a learner and is promoted to a voting member using the same learner-to-voter promotion mechanism that CEO uses during cluster bootstrapping.

The overall orchestration differs from bootstrapping: bootstrapping uses a temporary bootstrap member that is later removed before the cluster reaches steady state, while a Day 2 transition adds permanent members to a running production cluster. Critically, the 2-voter intermediate state (steps 4–5 below) is unique to Day 2 transitions — it does not occur during bootstrapping.

1. **Starting state**: 1 etcd voting member (quorum=1)

> 2. CEO adds an etcd learner on the second control-plane node

Reviewer 3: Are you sure of this? I thought on a SNO cluster (which is what it is at this point in the transition) it uses the unsafe scaling strategy which doesn't actually use learners. Have you verified this?

Or is this explaining a change that needs to be implemented? Might be good to clarify the wording somewhere to explain it's to be implemented if that's the case

Author: ack, I'm leaning on other peoples etcd knowledge for this so I'll ping @jaypoulz for more clarification

Author: This needs clarification from the etcd team. CEO's unsafe scaling strategy on SNO and whether it uses learners needs verification. The document now states this as a dependency requiring confirmation during dev preview.

Reviewer 7: The CEO uses learners regardless, the "unsafe" aspect is that the CEO doesn't check for quorum or member health on SNO, it just adds the new member as a learner and goes through the normal promotion process.

Reviewer 4: Etcd scales with learners. The strategies only determine when you are allowed to exit bootstrap mode.

#### Component Changes Summary

| Component | Changes Required |
| --------- | ---------------- |
| cluster-config-operator | New topology transition controller; watches `spec.desiredTopology`, coordinates transitions, updates status topology fields |
| Infrastructure API (`openshift/api`) | Add `desiredTopology` to `InfrastructureSpec`; update immutability documentation on status topology fields; ValidatingAdmissionPolicy for topology field edits |
| `oc` CLI | New `oc adm transition topology` command |
| cluster-etcd-operator | Sequential etcd scaling during transitions (learner-to-voter promotion mechanism from bootstrapping) |
| ingress, networking, monitoring operators | Reconcile on infrastructure status topology field changes |

#### Platform Support Constraints

See [Standalone Clusters](#standalone-clusters) for platform support details. The initial implementation targets `platform: none` only; `platform: baremetal` and cloud platforms are future work.

> The node-joining mechanism is initially UPI, but the design must not preclude future use of the Machine API Operator (MAO) for automated node provisioning. The topology transition controller checks for Node objects in the Kubernetes API regardless of how they were provisioned, so MAO-provisioned nodes are expected to work without controller changes. However, Machine object lifecycle implications have not yet been validated.

Reviewer 3: There's nothing about this EP that should care about how the Node came to be

Author: Ack, added this in response to previous review comments, I'll take another look at this

Author: Agreed — simplified to: 'The topology transition controller checks for Node objects in the API regardless of how they were provisioned.'

### Risks and Mitigations

#### Risk: Quorum Loss During Two-Member Transient State

**Risk**: During sequential etcd scaling (1→2→3), the cluster passes through a 2-member state where quorum=2. Losing either member during this window is fatal — the cluster loses its API and requires manual recovery.

**Mitigation**:
- The 2-member state is transient and the learner-to-voter promotion mechanism is reused from cluster bootstrapping — a well-exercised code path
- Learner instances are used before promoting members to minimize the promotion window
- No availability guarantee during transitions; administrators should treat scaling operations as a maintenance window

> - CEO will attempt rollback if scaling fails (e.g., rollback to 1 member if the 1→2→3 scale-up fails partway through)

Reviewer 3: The more I think about this, the less I think this is actually possible.

If you lose quorum, the API server will stop working. How is the etcd operator then going to continue to operate, understand that the quorum is lost, and then go and attempt to recover?

I'm pretty sure this has to be a manual intervention at this point no?

Author: I'm not sure the details around the quorum restore script and if it's automated or manually triggered. @jaypoulz @tjungblu do you have more detail I can add around this?

Author: Agreed — quorum loss during the 2-member window requires manual intervention via `quorum-restore.sh`. Automated recovery is not possible without quorum. Updated throughout the document to be consistent on this point.

- Future iterations may explore admitting two learners simultaneously and promoting only when both are ready, eliminating the 2-member voting window entirely, but that is out of scope for this enhancement

#### Risk: Transition Fails Partway Through

**Risk**: A transition may fail after some operators have begun reconfiguring but before the transition completes, leaving the cluster in an intermediate state. For example, etcd scales to 2 members, a network partition occurs between them, and both lose quorum with no API available.

**Mitigation**:
- The controller validates preconditions before starting

> - The controller sequences operations so that topology status fields are updated only after all setup steps succeed

Reviewer 3: What setup steps?

Author: Replaced vague 'setup steps' with specific references to etcd scaling in the risks section.

- Operators do not see a topology change until the controller updates the infrastructure status as the final step
- If setup steps fail, the controller reports the failure and CEO attempts rollback for etcd
- infrastructure status conditions provide detailed state for troubleshooting

#### Risk: Platform Bare Metal May Not Support Single-Node Clusters

**Risk**: If keepalived networking cannot be enabled, `platform: baremetal` will be limited to 2+ nodes, reducing the value of mutable topology for this platform.

**Mitigation**:
- Early coordination with the Bare Metal Networking team
- `platform: none` provides full support as a fallback
- The limitation can be documented while bare metal support is resolved

#### Risk: Cannot Validate External Requirements

> **Risk**: On `platform: none`, the topology transition controller cannot validate external requirements such as correct load balancer configuration or DNS setup. An administrator may initiate a transition with misconfigured networking, leading to a partially functional cluster.

Reviewer 3: This is the first time load balancers are mentioned. Is this still something we expect the CCO to validate? Feels like that's up to the admin to set up before they initiate the transition, and not something we should be caring about IMO

Author: ack, I'll wait for others to give thoughts on this before just agreeing. I'm torn between wanting to keep users from shooting themselves in the foot and just letting them do so.

Author: Agreed — LB configuration is the administrator's responsibility on `platform: none`. The 'Cannot Validate External Requirements' risk section documents this explicitly. The controller does not validate external infrastructure.

- A CLI process cannot provide persistent state tracking. A dropped SSH session or terminal close would leave the cluster in an intermediate state with no automated recovery.
- Error recovery and retry logic is better suited to a controller's reconciliation loop than imperative CLI code
- The CLI would need direct access to operator internals, violating separation of concerns

### Dedicated Topology Transition Operator

An earlier revision of this enhancement proposed a standalone topology transition operator deployed on-demand (not installed by default). The operator would own a transition CRD, manage the transition graph, and orchestrate the full transition lifecycle independently.

**Why it was rejected**:
- The scope does not warrant a new operator — cluster-config-operator is the natural home for this logic since it already owns the `config.openshift.io` API group and infrastructure CR lifecycle
- A standalone operator adds payload size, requires its own upgrade/lifecycle management, and introduces another component to monitor

> - The transition controller can live in CCO with zero overhead when not in use, gated by the `MutableTopology` feature gate

Reviewer 3: Nit: near zero

Author: Fixed — changed to 'near-zero overhead'.

#### Controller in machine-config-operator (MCO)

MCO handles node-level changes and rolling operations, making it a candidate for orchestrating node-topology changes.

**Why it was rejected**:
- MCO's domain is machine configuration (OS, kubelet config, node-level state), not cluster topology orchestration
- Topology transitions require cross-operator coordination (etcd, ingress, networking, infrastructure CR) that is outside MCO's current scope
- Like CVO and CEO, MCO is a critical-path operator where additional surface area increases risk to every cluster

**Note on CCO scope expansion**: The scope-expansion concern raised against CEO and MCO also applies to CCO, which currently focuses on CRD manifests and config synchronization. However, CCO is the canonical owner of the infrastructure CR and the `config.openshift.io` API group, making it the most natural home. The transition controller is also feature-gated with zero overhead when inactive, unlike CEO or MCO where additional code paths could affect core operations regardless of whether transitions are used.

## Open Questions

> 1. **HyperShift considerations**: Since the scope has broadened from edge-specific deployments to changing the topology assumption for OpenShift as a whole, do we need to consider HyperShift support? Initial answer is no — this would be future work and require its own enhancement.

Reviewer 3: This doesn't feel like an open question if it has an answer

Author: Yup, I'll move it

Author: Moved — HyperShift is already covered in the Topology Considerations section. Removed from Open Questions.

5. **Minimum resource requirements**: The controller should validate that new control-plane nodes meet minimum resource requirements before initiating a transition. The specific resource thresholds need to be defined.

## Test Plan

### CI Lanes

| Lane | Frequency | Description |
| ---- | --------- | ----------- |
| MutableTopology transition suite | Nightly | Run transition test suite: SNO → HA compact on `platform: none` |
| End-to-End tests (e2e) | Weekly | Standard test suite (openshift/conformance/parallel) on post-transition clusters |
| Upgrade between z-streams | Weekly | Test upgrades on post-transition clusters |
| Upgrade between y-streams | Weekly | Test upgrades across minor versions on post-transition clusters |

### CI Tests

#### Pre-Transition Tests

| Test | Description |
| ---- | ----------- |
| Precondition validation | Verify controller rejects transitions with missing nodes, invalid platforms, or unsupported source topologies |
| CLI interaction | Verify `oc adm transition topology` correctly patches `spec.desiredTopology` and monitors progress |

> | Feature gate gating | Verify the controller is inactive when `MutableTopology` feature gate is disabled |

Reviewer 3: The API won't exist when the gate is disabled, so you won't be able to drive the controller even if it were running. I think this test is probably impossible if not superfluous

Author: Agreed — removed the test. When the gate is disabled, the API doesn't exist and the controller isn't registered.

- CLI (`oc adm transition topology` command)
- Supported transition definitions and validation logic
- Infrastructure CR ValidatingAdmissionPolicy

**Control Plane Team:**
- cluster-etcd-operator (CEO) etcd scaling coordination

**Bare Metal Networking Team:**
- Bare metal networking for SNO clusters (future platform support)

**Component Teams:**
- Validate operator behavior during and after transitions

### Detecting Issues

**Transition Stuck or Failed:**
- Symptom: infrastructure status conditions show transition in progress or failed for an extended period
- Check: `oc get infrastructure cluster -o yaml` for status conditions
- Check: cluster-config-operator logs for transition controller errors
- Check: CEO logs for etcd scaling operations
- Resolution: Address the reported issue and retry, or contact support

**etcd Scaling Failures:**
- Symptom: etcd cluster unhealthy after transition attempt
- Check: CEO logs for etcd scaling operations
- Check: etcd member list: `oc -n openshift-etcd exec <etcd-pod> -- etcdctl member list`
- Resolution: CEO should attempt automatic rollback. If rollback fails, follow standard etcd disaster recovery procedures.

### Recovery Procedures

> 

Reviewer 8: As part of this transition, will backups scale ? 

If I take a backup on  SNO, wil it work on TNA or do I need to take a `fresh/new` backup ?

Author: I haven't thought through backups. @jaypoulz have you given this any thought? My initial thought is you would need to do a new backup as I'm not sure of how we would scale the backup.

Author: Good question — added as open question #5. A new backup should be taken after a successful transition. The interaction between pre-transition backups and post-transition cluster state needs investigation.

Reviewer 7: Using the pre-transition backup after the transition completes _should_ effectively revert the transition back to a SNO cluster. But +1 on this being tested.

Reviewer 4: The backup would have the old infrastructure state. I think testing this and providing a rollback procedure for bad transitions is probably a good target to define for a GA for this feature.

Mutable topology takes a different approach: instead of adding a new topology mode that operators must interpret, transitions are orchestrated by a controller in cluster-config-operator that coordinates the sequencing, validates preconditions, and updates the infrastructure CR only when the cluster is ready for the new mode. Operators continue to react to the same fixed topology values they already understand. This keeps operator logic simple and concentrates transition complexity in an existing core component.

### User Stories

* As a cluster administrator running Single Node OpenShift (SNO) at an edge location, I want to add control-plane nodes to my cluster to achieve high availability so that I can handle node failures without service disruption as workloads become more critical.

* As a cluster administrator deploying OpenShift clusters at scale, I want to start with minimal footprint deployments that can grow into highly available clusters so that I can reduce initial costs while maintaining scalability.

> * As a cluster administrator managing a fleet of edge deployments, I want a supported path to transition my cluster topology so that I don't need to redeploy clusters when my infrastructure requirements change.

Reviewer 8: Do all of the edge deployment devices need to be identical ? With as fast as hardware is moving, this might present a challenge as new hardware is added.

Author: No, we don't do anything in terms of checking hardware. For this first transition set (SNO->HA) it's primarily that the nodes are there, reachable, and have etcd running on them

#### Single-node Deployments or MicroShift

Single Node OpenShift (SNO) clusters are the primary source topology for transitions. The initial use case is enabling SNO deployments to transition to HA compact (3-node) configurations as requirements change.

The topology transition controller is gated by the `MutableTopology` feature gate and has no resource impact on clusters that do not use this feature.

Image Based Install (IBI) clusters are out of scope for this enhancement. Whether IBI clusters can support topology transitions has not been evaluated.

MicroShift is not affected by this enhancement and is unlikely to be included as a supported transition target.

#### OpenShift Kubernetes Engine

This proposal does not depend on features excluded from the OpenShift Kubernetes Engine (OKE) product offering. Mutable topology modifies core infrastructure components — the infrastructure API, cluster-config-operator, cluster-etcd-operator, and other in-payload operators — all of which are included in OKE.

### Workflow Description

#### Transition: SNO to HA Compact (3-Node)

> **Operational guidance**: Administrators should treat topology transitions as a maintenance window. Cluster availability is not guaranteed during the transition — particularly during the 2-member etcd window where any control-plane node failure is fatal. Administrators should reduce non-critical workload risk accordingly. Administrators should take an etcd backup before rand after a successful transition (see [Open Questions](#open-questions) regarding pre-transition backup compatibility).

Reviewer 8: Is there a typo in the before `and`  after?

```suggestion
**Operational guidance**: Administrators should treat topology transitions as a maintenance window. Cluster availability is not guaranteed during the transition — particularly during the 2-member etcd window where any control-plane node failure is fatal. Administrators should reduce non-critical workload risk accordingly. Administrators should take an etcd backup before and after a successful transition (see [Open Questions](#open-questions) regarding pre-transition backup compatibility).
```

**Spec (user intent):**

A new `controlPlaneTopology` field is added to `InfrastructureSpec` to express the administrator's intent to transition:

```go
// DesiredControlPlaneTopologyMode restricts the set of topology modes that can be
// requested as a transition target.
// +kubebuilder:validation:Enum=SingleReplica;HighlyAvailable
type DesiredControlPlaneTopologyMode string

const (
	DesiredSingleReplica   DesiredControlPlaneTopologyMode = "SingleReplica"
	DesiredHighlyAvailable DesiredControlPlaneTopologyMode = "HighlyAvailable"
)

type InfrastructureSpec struct {
	CloudConfig  ConfigMapFileReference `json:"cloudConfig"`
	PlatformSpec PlatformSpec           `json:"platformSpec,omitempty"`
	// ControlPlaneTopology expresses the administrator's intent
	// for the cluster's control plane topology. Empty by default — the
	// field is unset until an administrator explicitly initiates a
	// transition. When set and the value differs from
	// status.controlPlaneTopology, the topology transition controller
	// in cluster-config-operator initiates a transition. An empty value
	// means no transition has been requested.
	// +optional
	// +openshift:enable:FeatureGate=MutableTopology

> 	ControlPlaneTopology DesiredControlPlaneTopologyMode `json:"controlPlaneTopology,omitempty"`

Reviewer 9: Have you considered defining enums for the exact supported transitions?

For example

```go
DesiredSingleReplicaToHighlyAvailable DesiredControlPlaneTopologyModeTransition = "SingleReplicaToHighlyAvailable"
```

The advantage is forward support for the clients.

The current `Single -> HA` direction is only present in the validation. The client doesn't know that `HA -> Single` cannot occur. It either has to support both from the start, or it will break once we add this transition in future release.

With the exact enum, the client can do

```go
switch topologyTransition {
case "", "SingleReplicaToHighlyAvailable":
  scaleUp() // no need for scale down logic
default
  panic("not supported") // or report Degraded condition
}
```

Is this enum also customer facing?

If we add the above logic to our operators, it would be easy to see in the CI who gets broken once we add new states and transitions.

Author: Rather than an enum, we are planning an expansion to infra status that operates in a similar manner to upgrades. The goal is add a status field that specifies the valid transitions (to be evaluated periodically by the CCO controller). This will make the logic in the cli much simpler and instead of duplicating the logic that exists in the controller to protect transitions, it can check the status field for what's valid and fail fast. My opinion is that's implementation detail beyond this enhancement though. For now, adopting the standard spec/status relationship is in my opinion the right way to do this.

Reviewer 9: Sounds good. @jeff-roche could you please also mention this in the enhancement as a followup? It would be good to mention this in the graduation criteria as well - to at least revisit the topic during graduation.

- The limitation can be documented while bare metal support is resolved

#### Risk: Cannot Validate External Requirements

**Risk**: On `platform: none`, the topology transition controller cannot validate external requirements such as correct load balancer configuration or DNS setup. An administrator may initiate a transition with misconfigured networking, leading to a partially functional cluster.

**Mitigation**:
- Pre-flight checks validate what is within the cluster's control (node presence, resource requirements, operator health)
- External requirements (VIPs, DNS, load balancer configuration) are documented as the administrator's responsibility
- The CLI can surface warnings about external prerequisites before patching the infrastructure CR

### Drawbacks

#### Coordination Across Teams

The SNO-to-HA transition requires coordination with CEO, ingress, networking, and other operator teams to ensure they reconcile correctly when topology status fields change. This is less coordination than the previous Adaptable Topology approach (which required every operator to handle dynamic node-count awareness), but still significant.

#### OLM Operators and Topology Changes

> OLM-managed operators that read topology values at startup (rather than watching for changes) will not automatically react to topology transitions. These operators will need to either be updated to watch the infrastructure CR for topology changes, or be restarted after a transition completes. The scope of affected operators needs investigation.

Reviewer 9: - Managing the list of operators that need to be restarted is difficult. From maintainability perspective, it is much easier to teach these operators to watch for these transition changes.
- Once we implement the scale down direction, this can cause stability issues or degradation if the operators are not expecting it.
- One issue is, that to add a new transition, every operator has to be updated. Or it has to support every transition in the beginning to future proof it. None of these are very client friendly.

Reviewer 4: I feel like we could have a whole EP dedicated to management of OLM operators during topology transitions. I do think that being infrastructure-aware is up to the operator, but currently we don't have a way of know if an operator supports at topology. In fact, the only aspect of this we do support is architecture and os compatibility.

I would like to see us propose a manifest enhancement to OLM operators that allows them to specificy which topologies they support (or don't). From there we're operating optimistically when data is missing, but we can be much more confidant when it's present. Additionally, we can make this a requirement for catalog operators, so transitions can happen with confidence for operators we publish. Right now, we don't have anything like this - operators are already expected to support all topologies by default, even as we add new ones.

Reviewer 9: Can we also track this in the enhancement?

Reviewer 4: @jeff-roche Can we be more explicit about future extensions of optional-operator based transition declarations being the subject of a future enhancement. More specifically, the scope of this feature - pre-GA will be on the core operators? Before we GA this feature we should have at least a plan defined for how we want to improve the UX around optional operators. But actually managing them is beyond the scope of this feature.

* Maintain backward compatibility — existing clusters with fixed topology modes are unaffected
* Establish the architectural foundation for additional transition paths in the future

### Non-Goals

* Supporting all possible topology transitions in the initial implementation (only SNO → HA compact on `platform: none`)
* Supporting transitions for HyperShift or hosted control plane clusters
* Supporting transitions for MicroShift deployments
* Supporting transitions for Image Based Install (IBI) clusters
* Automatic node provisioning or deprovisioning based on workload demands
* Scaling down control-plane or worker nodes (scale-down may be addressed in a future enhancement)
* Supporting bidirectional transitions (e.g., HA → SNO) in the initial implementation

## Proposal

This enhancement introduces a new infrastructure API field and a topology transition controller in cluster-config-operator (CCO; not to be confused with cloud-credential-operator) to enable topology transitions as Day 2 operations.

The approach follows the standard OpenShift spec/status contract and mirrors the pattern used by `oc adm upgrade`:

> 1. **`desiredControlPlaneTopology` field in InfrastructureSpec** — Expresses the administrator's intent to transition. The CLI patches this field to initiate a transition. The existing `controlPlaneTopology` and `infrastructureTopology` fields in status continue to represent the cluster's observed topology.

Reviewer 3: Nit, I don't think we need `desired` in the name here. Kubernetes is declarative, every spec field is the "desired" state that operators should reconcile towards

2. **Topology transition controller in cluster-config-operator** — A new controller in CCO that watches the infrastructure CR for `desiredControlPlaneTopology` changes, validates preconditions, coordinates the transition, and updates the status topology fields when the cluster is ready for the new mode.

> 3. **`oc adm transition topology` CLI command** — A command that validates preconditions, patches `spec.desiredControlPlaneTopology` on the infrastructure CR, and returns immediately.

Reviewer 10: As of now this is mainly focused on control plane topology, but as this becomes more developed would we want to have this be under an umbrella term to allow easier additions and more logical groupings, wdyt?

```yaml
spec:
   topology:
        desiredControlPlaneTopology: "HighlyAvailable"
        desiredInfraTopology: "Single" # Possible future things
```

Author: Would we need desired or would it being located under spec imply desired? I'm wondering if we could just match the status field names and let spec imply desired. In terms of adding it to a topology object, I'm fine with it but I'm also curious @JoelSpeed's thoughts

Reviewer 3: > we could just match the status field names and let spec imply desired.

100% do this

#### Single-node Deployments or MicroShift

Single Node OpenShift (SNO) clusters are the primary source topology for transitions. The initial use case is enabling SNO deployments to transition to HA compact (3-node) configurations as requirements change.

The topology transition controller is gated by the `MutableTopology` feature gate and has no resource impact on clusters that do not use this feature.

Image Based Install (IBI) clusters are out of scope for this enhancement. Whether IBI clusters can support topology transitions has not been evaluated.

MicroShift is not affected by this enhancement and is unlikely to be included as a supported transition target.

#### OpenShift Kubernetes Engine

This proposal does not depend on features excluded from the OpenShift Kubernetes Engine (OKE) product offering. Mutable topology modifies core infrastructure components — the infrastructure API, cluster-config-operator, cluster-etcd-operator, and other in-payload operators — all of which are included in OKE.

### Workflow Description

#### Transition: SNO to HA Compact (3-Node)

> **Operational guidance**: Administrators should treat topology transitions as a maintenance window. Cluster availability is not guaranteed during the transition — particularly during the 2-member etcd window where any control-plane node failure is fatal. Administrators should reduce non-critical workload risk accordingly. Administrators should take an etcd backup after a successful transition (see [Open Questions](#open-questions) regarding pre-transition backup compatibility).

Reviewer 7: I'd prioritize a backup before the transition, that way the admin can restore the cluster to a known working state if the scale fails.

##### Pre-Transition

> 1. The cluster administrator prepares at least 2 additional control-plane nodes and joins them to the cluster — the kubelet is running on each node and Node objects exist in the Kubernetes API. On `platform: none`, the administrator manages their own load balancing configuration (VIPs, DNS).

Reviewer 3: Nit, we only support 1 or 3 nodes, so this should be exactly 2 rather than at least 2

> 2. The cluster administrator runs `oc adm transition topology HighlyAvailable`

Reviewer 3: Have you had anyone responsible for `oc` review this EP yet?

Author: Not yet, do you have any suggestions?

Reviewer 3: @ardaguclu Are you able to review for CLI perspective? Or delegate?

Reviewer 4: https://redhat-internal.slack.com/archives/CB48XQ4KZ/p1784038299352329?thread_ts=1783456156.426299&cid=CB48XQ4KZ

Author: More directly, @ardaguclu has responded to the implementation PR we are using for the PoC: https://github.com/openshift/oc/pull/2295. 

The primary concern (non-blocking) was how nested the command is which is valid. I think for now we can leave it as is and look into moving the command up/simplifying it as we move towards tech preview and implementing more transition types (5.1 scope)

3. The CLI validates preconditions before patching (e.g., feature gate enabled, no transition already in progress)
4. The CLI patches the infrastructure CR: `spec.desiredControlPlaneTopology: HighlyAvailable`
5. The API server validates `desiredControlPlaneTopology` against the `DesiredTopologyMode` enum, rejecting unsupported topology modes before accepting the write

##### During Transition

6. The topology transition controller in CCO detects the `desiredControlPlaneTopology` change and validates preconditions:
   - Current `controlPlaneTopology` is `SingleReplica`
   - Target topology is `HighlyAvailable`
   - At least 3 nodes with `node-role.kubernetes.io/control-plane` or `node-role.kubernetes.io/master` labels are present in the Node API
   - No dedicated worker nodes are present (the initial implementation targets compact clusters only; clusters with dedicated workers require a different `infrastructureTopology` mapping that is not yet supported)
   - Platform is supported (`platform: none` in the initial implementation)

> 7. The controller signals that a transition is in progress (via CCO ClusterOperator status conditions)

Reviewer 3: Where do we define how it knows when the transition is in progress vs complete? 

The transition controller has one action, update the infrastructure status. What would it monitor after this point?

Author: I think it primarily comes down to monitoring the core payload operators we define as critical for this transition to make sure they reconcile the scaling without degrading

Reviewer 4: Yeah MCO renders new master config, ingress has 2 replicas, etcd has the right membership count, and no more SNO forced dnsmasq on the cluster = Done.

We do most of these checks preflight, but ensuring we're stable in the new topology gives us a sense that the transition was successful.

8. **Node-driven operator reactions** — these happen independently as new control-plane nodes join, before the topology status fields are updated:

>    - cluster-etcd-operator (CEO) detects new control-plane nodes and scales etcd members sequentially (1→2→3), reusing the learner-to-voter promotion mechanism from bootstrapping — each new node joins as a learner and is promoted to a voting member before the next is added

Reviewer 3: Do we want to teach etcd operator not to promote additional learners that is sees when the topology is SingleReplica? Why does it scale up unsafely today?

Have you had any etcd operator SMEs review this EP yet?

Reviewer 7: There's always an "unsafe" aspect to the scaling procedure as there's no way to promote multiple learners to voters in a single step, it always takes two steps to promote two members. During bootstrap of a typical 3 node cluster, if there's an issue with scaling, we can just redeploy. If, for some reason, the scale/promotion of the 3rd member fails in this case, that doesn't immediately put us in hot water, we can operate with 2 nodes as long as they agree and we don't lose another node. It gives some time for the admin to intervene (manually scaling or quorum-recovery or at worst restoring from backup or some other action).

Reusing the bootstrap process was a suggestion from the previous enhancement as it is a well-trod path since every 3 node cluster goes through the process; and I'm sure we'd hear about it if it was failing more often than not. Also, the CEO is designed to handle this process so support for this feature would essentially be free by re-using the scaling process.

   - `infrastructureTopology` transitions from `SingleReplica` to `HighlyAvailable` (no dedicated workers, so it matches control plane topology)
   - `mastersSchedulable` is set based on the transition definition (remains `true` for SNO → HA compact)
10. **Topology-driven operator reactions** — operators that watch the infrastructure status topology fields reconcile against the new values and adjust their deployment strategies, replica counts, and placement policies. This is a distinct phase from step 8: step 8 covers operators reacting to node presence, step 10 covers operators reacting to the topology status change. The set of operators with topology-dependent behavior has not been fully enumerated — building the per-operator topology dependency matrix is a prerequisite for entering dev preview (see [Graduation Criteria](#entering-dev-preview) and [Open Questions](#open-questions)).

    **Note**: OLM-managed operators that read topology at startup rather than watching for changes may need to be restarted after the transition completes. See [Drawbacks](#drawbacks) for details.

##### Post-Transition

> 11. The controller validates that critical operators have reconciled to a healthy state

Reviewer 3: How do you know that the operators are reacting to the topology change vs any other observation they have made within the cluster?

Author: I think primarily replica count. I think we need to identify for this transition what core operators we need to wait for and what the change looks like for each of them so that we can validate the critical pieces have transitioned

Reviewer 4: - etcd/apiserver react to the control-plane node count and scale accordingly
- etcd specifically has topology specific criteria for when it's allowed to exit bootstrap mode; after bootstrap it doesn't care (except for the TNF specific code which is gated by topology)
- console reads topology at creation time and injects in window.SERVER_FLAGS to customize some display and plugin behavior
- ingress hardcodes 1 replica if you're SNO at ingresscontroller creation time 😠 This is a bug IMHO, but it's done like this because of limitation when how k8s handles restarts IIRC; to fix it you delete the ingress operator pod and the ingresscontroller to force it to reevaludate the defaults.
- MCO rerenders the master-configs when topology changes (this is the big one)

How do you know the operators are reacting to the topology change? You don't really, but you do know the outcomes of what defines the runtime criteria of the target topology. we can introduce post-transition validations that define the exit criteria of the transition.

Author: Currently we implement a soak period that allows operators to react. Some naturally do and getting the status of the cluster operators is the easy path currently. I think as we move towards tech preview, implement more testing and find edge cases we will want to expand this. For now though, I think just monitoring the cluster operator list is good enough to report on.

12. The controller sets the `TopologyTransitionCompleted` condition on the CCO ClusterOperator status. The infrastructure status reflects the completed transition — `desiredControlPlaneTopology` matches `status.controlPlaneTopology`, so no further action is taken.

The CLI returns immediately after patching `desiredControlPlaneTopology` (step 4). Administrators can monitor transition progress by watching CCO ClusterOperator status conditions (e.g., `oc get clusteroperator cluster-config-operator -o yaml`).

##### Failure Handling

If a transition fails partway through:

- The controller sets a `TopologyTransitionFailed` condition on the CCO `ClusterOperator` status with diagnostic information

> - For etcd scaling failures during the 2-member window, quorum is lost and automated recovery is not possible — the administrator must manually run `quorum-restore.sh` per standard etcd disaster recovery procedures

Reviewer 3: During a quorum loss scenario, writes to the cluster are paused, you won't actually be able to write a Failed condition status

If this is the only failure we can identify, I don't think a failed condition is actually going to be possible to implement

Reviewer 7: A note here about wording, if we're at 2 voting members and a learner we can't promote, we do have quorum. The failure scenario would be being at 2 voting members, failing to scale to a 3rd, then losing one of the voting members; that would lose us quorum and quorum-recovery would be necessary.

Author: > The failure scenario would be being at 2 voting members, failing to scale to a 3rd

@dusk125 in that scenario, would we be able to report a failure or would writes be paused?

Author: @dusk125 I misunderstood the first time I read this. We would be able to report that a learner was not promoted to a voting member, we would not be able to report failure in the event that a voting member goes down. If a voting member goes down, how does the admin know? (I'm assuming this is documented somewhere)

Reviewer 7: Yeah, if you lose quorum, you can't write. What do you mean by "reporting a failure"?

Author: There will be a new status field in CCO reflecting transition status, the plan is to have the controller monitor the etcd transition to validate it completed successfully and report success/failure. The target is to utilize that status field as the source of truth for if the transition completed or failed

#### Controller in machine-config-operator (MCO)

MCO handles node-level changes and rolling operations, making it a candidate for orchestrating node-topology changes.

**Why it was rejected**:
- MCO's domain is machine configuration (OS, kubelet config, node-level state), not cluster topology orchestration
- Topology transitions require cross-operator coordination (etcd, ingress, networking, infrastructure CR) that is outside MCO's current scope
- Like CVO and CEO, MCO is a critical-path operator where additional surface area increases risk to every cluster

**Note on CCO scope expansion**: The scope-expansion concern raised against CEO and MCO also applies to CCO, which currently focuses on CRD manifests and config synchronization. However, CCO is the canonical owner of the infrastructure CR and the `config.openshift.io` API group, making it the most natural home. The transition controller is also feature-gated with near-zero overhead when inactive, unlike CEO or MCO where additional code paths could affect core operations regardless of whether transitions are used.

## Open Questions

> 1. **Learner promotion after voter failure**: If CEO runs a learner on a second control-plane node and the voter fails, can quorum restore promote the learner? Or can only former voters be restored with quorum?

Reviewer 7: I don't think so, learners can't self promote and promoting requires quorum. This is effectively equivalent to losing your node in SNO - which is why it would be important to take a backup before initiating this process.

2. **OLM operator impact**: Which OLM-managed operators read topology values? Do they watch the infrastructure CR or read at startup only? This determines whether operators need code changes or just a restart after transition.

3. **Per-operator transition behavior**: The transition behavior for CEO is understood (etcd sequential scaling). The specific requirements for ingress, networking, monitoring, and other operators during a topology transition need validation during dev preview. The per-operator topology dependency matrix is a prerequisite for entering dev preview — see [Graduation Criteria](#entering-dev-preview).

4. **Minimum resource requirements**: The controller should validate that new control-plane nodes meet minimum resource requirements before initiating a transition. The specific resource thresholds need to be defined.

> 5. **Backup compatibility across topologies**: If an administrator takes an etcd backup on a SNO cluster and later transitions to HA, is the pre-transition backup usable for restore on the post-transition cluster? A new backup should be taken after a successful transition, but the interaction between pre-transition backups and post-transition cluster state needs investigation.

Reviewer 7: Restoring a backup travels the cluster back in time. So a backup taken before the transition would make the cluster into a SNO cluster again when restored. This is something that should be tested though.

The before-transition backup would be essential as another layer of protection against failed scaling (quorum-restore being the first line of defense).

- SLOs documented and validated: target transition duration (SNO → HA compact), success rate threshold, and maximum time in the 2-member etcd window
- Monitoring and telemetry for transition metrics: Prometheus metrics exposed (transition_started, transition_completed, transition_failed, transition_duration_seconds) with alerts defined for stuck transitions exceeding SLO thresholds
- Support procedures documented
- Feature gate moved to `Default` feature set

### Removing a deprecated feature

N/A

## Upgrade / Downgrade Strategy

### Upgrades

Clusters that have undergone topology transitions follow standard OpenShift upgrade procedures. The resulting topology values (`HighlyAvailable`, `SingleReplica`, etc.) are existing enum values that all operators already support. There are no special upgrade considerations for post-transition clusters.

The topology transition controller upgrades as part of cluster-config-operator via the standard CVO-managed upgrade path.

### Downgrades

**Z-stream downgrades** (within a minor version that supports mutable topology):
Standard downgrade procedures apply. Completed transitions are not reverted — the cluster retains its current topology.

> **Y-stream downgrades** (to a minor version without mutable topology support):

Reviewer 7: The CVO blocks Y-stream downgrades: https://github.com/openshift/cluster-version-operator/blob/4cfd94b468732a9fff4c29afa679085fcf363e5f/pkg/payload/precondition/clusterversion/rollback.go#L71-L110

4. The CLI patches the infrastructure CR: `spec.controlPlaneTopology: HighlyAvailable`
5. The API server validates `controlPlaneTopology` against the `DesiredControlPlaneTopologyMode` enum, rejecting unsupported topology modes before accepting the write

##### During Transition

6. The topology transition controller in CCO detects the `controlPlaneTopology` change and validates preconditions:
   - Current `status.controlPlaneTopology` is `SingleReplica`
   - Target topology is `HighlyAvailable`
   - At least 3 nodes with `node-role.kubernetes.io/control-plane` or `node-role.kubernetes.io/master` labels are present in the Node API
   - No dedicated worker nodes are present (the initial implementation targets compact clusters only; clusters with dedicated workers require a different `infrastructureTopology` mapping that is not yet supported)
   - Platform is supported (`platform: none` in the initial implementation)
7. The controller signals that a transition is in progress (via CCO ClusterOperator status conditions)
8. **Node-driven operator reactions** — these happen independently as new control-plane nodes join, before the topology status fields are updated:

>    - cluster-etcd-operator (CEO) detects new control-plane nodes and scales etcd members sequentially (1→2→3), reusing the learner-to-voter promotion mechanism from bootstrapping — each new node joins as a learner and is promoted to a voting member before the next is added. This is a thoroughly tested procedure that all 3 control plane node clusters go through.

Reviewer 4: Adding nodes isn't automated so it's safer to document the procedure to force users to do this before allowing them to admit the transition. This way, that procedure is not part of the scope of our feature, it's a prerequisite to use it. Please drop this point of feedback.

12. The controller sets the `TopologyTransitionCompleted` condition on the CCO ClusterOperator status. The infrastructure status reflects the completed transition — `spec.controlPlaneTopology` matches `status.controlPlaneTopology`, so no further action is taken.

The CLI returns immediately after patching `spec.controlPlaneTopology` (step 4). Administrators can monitor transition progress by watching CCO ClusterOperator status conditions (e.g., `oc get clusteroperator cluster-config-operator -o yaml`).

##### Failure Handling

If a transition fails partway through:

- The controller sets a `TopologyTransitionFailed` condition on the CCO `ClusterOperator` status with diagnostic information
- For etcd scaling failures during the 2-member window, if a voting member is lost, quorum is lost and automated recovery is not possible — the administrator must manually run `quorum-restore.sh` per standard etcd disaster recovery procedures. In this case, the controller would not be able to set the TopologyTransitionFailed status due to writes being paused on the cluster during quorum loss.
- The administrator can inspect CCO logs and ClusterOperator status conditions for details

> - `spec.controlPlaneTopology` remains unchanged — the controller will retry the transition with exponential backoff when preconditions are met. To cancel a transition that has not yet reached the status update (step 9), the administrator resets `spec.controlPlaneTopology` to match the current `status.controlPlaneTopology` (e.g., `oc adm transition topology SingleReplica`). After the status fields have been updated, the transition is effectively complete and cannot be cancelled — the cluster is in the new topology. This follows the standard Kubernetes pattern where controllers continuously reconcile toward the desired state until the user changes intent

Reviewer 4: There's a status condition for transition in progress. Please close this.

> - `spec.controlPlaneTopology` remains unchanged — the controller will retry the transition with exponential backoff when preconditions are met. To cancel a transition that has not yet reached the status update (step 9), the administrator resets `spec.controlPlaneTopology` to match the current `status.controlPlaneTopology` (e.g., `oc adm transition topology SingleReplica`). After the status fields have been updated, the transition is effectively complete and cannot be cancelled — the cluster is in the new topology. This follows the standard Kubernetes pattern where controllers continuously reconcile toward the desired state until the user changes intent

Reviewer 4: please resolve

| Condition Type | Meaning |
| -------------- | ------- |
| `TopologyTransitionProgressing` | A transition is in progress. `status: True` when actively transitioning, `status: False` when idle or complete. |
| `TopologyTransitionCompleted` | The most recent transition completed successfully. `status: True` after successful completion. |
| `TopologyTransitionFailed` | The most recent transition attempt failed. `status: True` when a failure has occurred. `message` contains diagnostic details. |

These condition types provide a stable contract for the CLI, console, and telemetry consumers. Reason values (e.g., `TransitionStarted`, `EtcdScalingInProgress`, `WaitingForOperators`, `PreconditionNotMet`, `EtcdScalingFailed`) will be refined during dev preview implementation.

#### Admission Control

**Spec validation**: The `DesiredControlPlaneTopologyMode` named type restricts `spec.controlPlaneTopology` to the set of topology modes that have defined transitions (`SingleReplica`, `HighlyAvailable`). The API server rejects unsupported values at admission time via the kubebuilder enum validation on the type. No additional validation rules are required.

> Access to `spec.spec.controlPlaneTopology` is governed by the existing RBAC for the infrastructure CR (`infrastructures.config.openshift.io`). By default, only users with `cluster-admin` or equivalent roles can modify infrastructure spec fields. No additional RBAC restrictions are proposed for the initial implementation; a dedicated role for topology transitions may be considered in future iterations if finer-grained access control is needed.

Reviewer 4: This is a nit. If we need to do another round of revisions, we'll fix this, but this is not needed for this final review.

> Access to `spec.spec.controlPlaneTopology` is governed by the existing RBAC for the infrastructure CR (`infrastructures.config.openshift.io`). By default, only users with `cluster-admin` or equivalent roles can modify infrastructure spec fields. No additional RBAC restrictions are proposed for the initial implementation; a dedicated role for topology transitions may be considered in future iterations if finer-grained access control is needed.

Reviewer 4: no. that's not needed.

> Access to `spec.spec.controlPlaneTopology` is governed by the existing RBAC for the infrastructure CR (`infrastructures.config.openshift.io`). By default, only users with `cluster-admin` or equivalent roles can modify infrastructure spec fields. No additional RBAC restrictions are proposed for the initial implementation; a dedicated role for topology transitions may be considered in future iterations if finer-grained access control is needed.

Reviewer 4: please resolve

## Open Questions

2. **OLM operator impact**: Which OLM-managed operators read topology values? Do they watch the infrastructure CR or read at startup only? This determines whether operators need code changes or just a restart after transition.

3. **Per-operator transition behavior**: The transition behavior for CEO is understood (etcd sequential scaling). The specific requirements for ingress, networking, monitoring, and other operators during a topology transition need validation during dev preview. The per-operator topology dependency matrix is a prerequisite for entering dev preview — see [Graduation Criteria](#entering-dev-preview).

4. **Minimum resource requirements**: The controller should validate that new control-plane nodes meet minimum resource requirements before initiating a transition. The specific resource thresholds need to be defined.

> 5. **Backup compatibility across topologies**: If an administrator takes an etcd backup on a SNO cluster and later transitions to HA, is the pre-transition backup usable for restore on the post-transition cluster? A new backup should be taken after a successful transition, but the interaction between pre-transition backups and post-transition cluster state needs investigation. Ideally restoring the pre-transition backup would revert the cluster to SNO, but that flow needs to be validated.

Reviewer 4: A previously open question was retired. This is nit that's not worth fixing unless we need another revision.

> 5. **Backup compatibility across topologies**: If an administrator takes an etcd backup on a SNO cluster and later transitions to HA, is the pre-transition backup usable for restore on the post-transition cluster? A new backup should be taken after a successful transition, but the interaction between pre-transition backups and post-transition cluster state needs investigation. Ideally restoring the pre-transition backup would revert the cluster to SNO, but that flow needs to be validated.

Reviewer 4: please resolve

## Test Plan

### CI Lanes

| Lane | Frequency | Description |
| ---- | --------- | ----------- |
| MutableTopology transition suite | Nightly | Run transition test suite: SNO → HA compact on `platform: none` |
| End-to-End tests (e2e) | Weekly | Standard test suite (openshift/conformance/parallel) on post-transition clusters |
| Upgrade between z-streams | Weekly | Test upgrades on post-transition clusters |
| Upgrade between y-streams | Weekly | Test upgrades across minor versions on post-transition clusters |

### CI Tests

#### Pre-Transition Tests

| Test | Description |
| ---- | ----------- |
| Precondition validation | Verify controller rejects transitions with missing nodes, invalid platforms, or unsupported source topologies |
| CLI interaction | Verify `oc adm transition topology` correctly patches `spec.controlPlaneTopology` and monitors progress |

> 

Reviewer 4: The test monitors progress, not the command. You're reading the sentence wrong. It's ambiguously written, but this is not an issue.

- Topology transition controller in cluster-config-operator
- CLI (`oc adm transition topology` command)
- Supported transition definitions and validation logic
- Infrastructure CR API changes (`DesiredControlPlaneTopologyMode` type, `controlPlaneTopology` field)

**Control Plane Team:**
- cluster-etcd-operator (CEO) etcd scaling coordination

**Bare Metal Networking Team:**
- Bare metal networking for SNO clusters (future platform support)

**Component Teams:**
- Validate operator behavior during and after transitions

### Detecting Issues

**Transition Stuck or Failed:**
- Symptom: CCO ClusterOperator status conditions show transition in progress or failed for an extended period
- Check: `oc get clusteroperator cluster-config-operator -o yaml` for status conditions
- Check: cluster-config-operator logs for transition controller errors
- Check: CEO logs for etcd scaling operations
- Resolution: Address the reported issue and retry, or contact support

**etcd Scaling Failures:**
- Symptom: etcd cluster unhealthy after transition attempt
- Check: CEO logs for etcd scaling operations
- Check: etcd member list: `oc -n openshift-etcd exec <etcd-pod> -- etcdctl member list`

> - Resolution: If quorum is lost, follow standard etcd disaster recovery procedures (`quorum-restore.sh`). Automated rollback is not possible without quorum. Restoring to pre-transition snapshot could operate as a fallback recovery procedure pending verification of that procedure. 

Reviewer 4: In terms of restoration proceeds, this is nothing more tested in OCP than etcd backup restore. This is not a concern.

---
title: mutable-topology
authors:
  - "@jeff-roche"
  - "@jaypoulz"
  - "@eggfoobar"
reviewers:
  - "@tjungblu, for cluster-etcd-operator"
  - "@dusk125, for cluster-etcd-operator"
  - "@joelspeed, for API, infrastructure config, and cluster-config-operator scope"
  - "@patrickdillon, for OpenShift installer"
  - "@zaneb, for metal platform interaction"
  - "@ardaguclu, for OpenShift client"
  - "@atiratree, for OpenShift client"
  - "@jerpeter, for OpenShift architecture"
  - "@sdodson, for architecture"
  - "@dgoodwin, for architecture"
approvers:
  - "@jerpeter, for OpenShift architecture"
api-approvers:
  - "@joelspeed, for API and infrastructure config"
creation-date: 2026-05-11
last-updated: 2026-05-29
tracking-link:

>   - https://issues.redhat.com/browse/OCPEDGE-2280

Author: this is fine

6. The API server validates `controlPlaneTopology` against the `DesiredControlPlaneTopologyMode` enum, rejecting unsupported topology modes before accepting the write

##### During Transition

7. The topology transition controller in CCO detects the `controlPlaneTopology` change and validates preconditions:
   - Every ClusterOperator other than cluster-config-operator itself reports `Available=True`, `Progressing=False`, `Degraded=False`
   - Exactly 3 nodes with `node-role.kubernetes.io/control-plane` or `node-role.kubernetes.io/master` labels are present, all schedulable and `Ready`
   - No dedicated worker nodes are present (the initial implementation targets compact clusters only; clusters with dedicated workers require a different `infrastructureTopology` mapping that is not yet supported)
   - etcd already reports quorum, is not mid-scaling, and already has 3 voting members — i.e., step 2's node-driven scaling has already finished
   If any precondition fails — including an etcd that has not yet finished scaling — the controller does not admit the transition; it records the reason and re-evaluates on the next sync (see [Failure Handling](#failure-handling)).

> 8. Once preconditions pass, the controller sets `Upgradeable=False` and a `Progressing` condition on the CCO `ClusterOperator` in the same update, signaling that a transition is in progress and preventing CVO from initiating an upgrade

Author: This would need to be handled by the admin pre-transition, who likely will not trigger an upgrade while adding nodes. 
Won't do, please resolve.

ControlPlaneTopology TopologyMode `json:"controlPlaneTopology"`

// infrastructureTopology expresses the expectations for infrastructure
// services that do not run on control plane nodes. When there are no
// dedicated worker nodes, this is set to match controlPlaneTopology.
// +kubebuilder:default=HighlyAvailable
InfrastructureTopology TopologyMode `json:"infrastructureTopology,omitempty"`
```

No new enum values are added to `TopologyMode`. The existing values (`SingleReplica`, `HighlyAvailable`, `DualReplica`, `HighlyAvailableArbiter`) are sufficient.

**Transition progress** will be reported via the following condition types on the CCO `ClusterOperator` status:

| Condition Type | Meaning |
| -------------- | ------- |
| `TopologyTransitionControllerProgressing` | A transition has been admitted and post-transition validation has not yet passed. `status: True` while awaiting downstream reconciliation, `status: False` when idle, rejected, or complete. |
| `TopologyTransitionControllerUpgradeable` | Whether CVO may start an upgrade. `status: False` while a transition is requested, pending, or in progress; `status: True` when idle or complete. |

> These are existing operator conditions on the CCO `ClusterOperator`, not dedicated transition condition types — the `Reason`/`Message` fields distinguish states (e.g. `UnsupportedTransition`, `PreflightCheckFailed`, `TopologyTransitionInProgress`, `TopologyTransitionComplete`, `AsExpected`). There is no separate condition for etcd scaling — that scaling is a precondition the controller checks, not a state it tracks or reports on directly (see [Failure Handling](#failure-handling)). Reason values will be refined during dev preview implementation.

Author: implementation detail that is out of scope

2. Re-read the Infrastructure CR and confirm the requested spec has not changed since preconditions were checked
3. Update `status.controlPlaneTopology` and `status.infrastructureTopology` to the target value
4. On later syncs, wait a soak period (5 minutes) after the `Progressing` condition was set, then begin checking post-transition validation criteria

If no supported transition matches, or a precondition fails, the controller records the reason on the `Progressing`/`Upgradeable` conditions rather than erroring, so the administrator can revert `spec.controlPlaneTopology` to resolve it.

**Validation criteria** (checked once the soak period has elapsed; all must pass to consider the transition complete):

- 3 control-plane nodes remain schedulable and `Ready`, and a minimum of 2 worker-labeled nodes, including dual-role compact nodes, are `Ready`.
- etcd retains quorum, is not mid-scaling, and reports 3 voting members
- The SNO-only master MachineConfig has been removed and a new rendered master/worker MachineConfigs exist, with the master MachineConfigPool reporting 3 ready machines
- The default IngressController reports a minimum of 2 available router replicas

> - kube-apiserver reports status for 3 nodes and openshift-apiserver reports 3 ready replicas

Author: implementation detail, out of scope

* Scaling down control-plane or worker nodes (scale-down may be addressed in a future enhancement)
* Supporting bidirectional transitions (e.g., HA → SNO) in the initial implementation

## Proposal

This enhancement introduces a new infrastructure API field and a topology transition controller in cluster-config-operator (CCO; not to be confused with cloud-credential-operator) to enable topology transitions as Day 2 operations.

The approach follows the standard OpenShift spec/status contract and mirrors the pattern used by `oc adm upgrade`:

1. **`controlPlaneTopology` field in InfrastructureSpec** — Expresses the administrator's intent to transition. The CLI patches this field to initiate a transition. The existing `controlPlaneTopology` and `infrastructureTopology` fields in status continue to represent the cluster's observed topology.

2. **Topology transition controller in cluster-config-operator** — A new controller in CCO that watches the infrastructure CR for `controlPlaneTopology` spec changes, validates preconditions, coordinates the transition, and updates the status topology fields when the cluster is ready for the new mode.

> 3. **`oc adm transition topology` CLI command** — A command that validates preconditions, patches `spec.controlPlaneTopology` on the infrastructure CR, and returns immediately.

Author: Operators need the change to trigger the transition. They won't react if we don't change it. 
Resolve this.

Access to `spec.controlPlaneTopology` is governed by the existing RBAC for the infrastructure CR (`infrastructures.config.openshift.io`). By default, only users with `cluster-admin` or equivalent roles can modify infrastructure spec fields. No additional RBAC restrictions are proposed for the initial implementation; a dedicated role for topology transitions may be considered in future iterations if finer-grained access control is needed.

**Status fields**: The existing topology status fields (`controlPlaneTopology`, `infrastructureTopology`, `mastersSchedulable`) are not protected by admission policies. This is consistent with other infrastructure status fields — no special protection exists for them today. An administrator who deliberately modifies these values outside the transition controller does so at their own risk.

#### Feature Gate

A new feature gate `MutableTopology` will be added to gate this functionality. The feature gate will progress through the following stages:

- **Dev Preview**: Part of the `DevPreviewNoUpgrade` feature set
- **Tech Preview**: Moved to the `TechPreviewNoUpgrade` feature set

> - **GA**: Moved to the `Default` feature set

Author: This is as it should be, default feature set will be tested before GA. 
Resolve this.

- Reports transition progress via CCO ClusterOperator status conditions

##### Supported Transitions

For the initial implementation:

```text
SingleReplica (SNO, platform: none) → HighlyAvailable (3-node compact)
```

Future transitions can be added without modifying the core controller logic. Each supported transition defines:

- **Preconditions**: What must be true before the transition can start
- **Orchestration steps**: What the controller coordinates during the transition
- **Validation criteria**: What must be true after the transition for it to be considered complete

##### Transition Orchestration

The controller reconciles `spec.controlPlaneTopology` against `status.controlPlaneTopology` on every sync. For the SNO → HA compact transition:

**Preconditions** (all must hold before a transition is accepted):

- Every ClusterOperator other than cluster-config-operator itself reports `Available=True`, `Progressing=False`, `Degraded=False`
- Exactly 3 control-plane-labeled nodes exist, all schedulable and `Ready`, and no dedicated worker nodes are present
- etcd has quorum and is not mid-scaling, and 3 voting members are already recorded for it

> These preconditions mean the administrator's node join and CEO's existing node-driven etcd scaling must already be complete — the controller does not itself trigger or wait on etcd scaling as part of the transition; it only accepts the transition once that has already happened.

Author: There are assumed basic limitations that are tested by the code, these are them. Resolving.

**Orchestration steps** (once preconditions pass):

1. Set `Upgradeable=False` and a `Progressing` condition on the CCO `ClusterOperator` in the same update, so CVO cannot start an upgrade while the transition is applied
2. Re-read the Infrastructure CR and confirm the requested spec has not changed since preconditions were checked
3. Update `status.controlPlaneTopology` and `status.infrastructureTopology` to the target value

> 4. On later syncs, wait a soak period (5 minutes) after the `Progressing` condition was set, then begin checking post-transition validation criteria

Author: This case is handled by the code flow, resolving.

- **GA**: Moved to the `Default` feature set

### Implementation Details/Notes/Constraints

#### Topology Transition Controller

A new topology transition controller is added to cluster-config-operator with the following characteristics:

- Watches the infrastructure CR for `spec.controlPlaneTopology` diverging from `status.controlPlaneTopology`
- Gated by the `MutableTopology` feature gate — inactive when the gate is disabled
- Maintains the set of supported transitions (initially only SingleReplica → HighlyAvailable on `platform: none`)
- Validates preconditions before starting a transition
- Updates `controlPlaneTopology` and `infrastructureTopology` in status once preconditions pass
- Reports transition progress via CCO ClusterOperator status conditions

##### Supported Transitions

For the initial implementation:

```text
SingleReplica (SNO, platform: none) → HighlyAvailable (3-node compact)
```

Future transitions can be added without modifying the core controller logic. Each supported transition defines:

- **Preconditions**: What must be true before the transition can start
- **Orchestration steps**: What the controller coordinates during the transition

> - **Validation criteria**: What must be true after the transition for it to be considered complete

Reviewer 9: How much do we expect for these to differ for each new transition?

It would be good to make these steps reusable/generic as much as we can. And parametrize the number of nodes per example. It might be worthwhile to consider adding such information (e.g. the number of masters/workers) to the API, to prevent an explosion of configurations an transitions.

Author: There is a core set that we think will be reused. The implementation PR currently does this in a reusable way so it's been accounted for. See https://github.com/openshift/cluster-config-operator/pull/495/changes#diff-973e21a8bf09a2990c12e5d6cf40bdb4e4e3a3a168074247ca21346d0825830dR39-R83 for details

Reviewer 9: The implementation bits look reasonable. cc @JoelSpeed in case we would like to do something with the API in that regard. Or maybe have a compatible way of adding that later?

- `DesiredControlPlaneTopologyMode` named type validated in API integration tests
- Per-operator topology dependency matrix completed: for each in-payload operator that reads `controlPlaneTopology` or `infrastructureTopology`, document what the operator uses the value for (replica count, scheduling, feature enablement) and whether it watches the infrastructure CR for changes or reads the value only at startup
- Operators that read topology only at startup are identified and a restart strategy is documented for post-transition reconciliation
- CCO sets `Upgradeable=False` on its ClusterOperator while a topology transition is in progress
- CI lanes operational for transition testing
- Developer documentation available

### Dev Preview -> Tech Preview

- Transition test suite validates full SNO → HA compact path
- Tests verify operator health during and after transitions
- Controller failure handling validated; etcd disaster recovery procedures documented for quorum loss scenarios
- `oc adm transition topology` command provides clear diagnostics on failure
- User-facing documentation in [openshift-docs](https://github.com/openshift/openshift-docs/)

> - End-to-end validation that CLI correctly patches `controlPlaneTopology` and the controller rejects unsupported transitions

Reviewer 9: Can we also revisit adding the supported transitions to the status during the TP graduation? Ref: https://github.com/openshift/enhancements/pull/2008#discussion_r3604276196

Author: Yeah, that's one of the major pieces we are tackling next so it make sense to add it as graduation criteria. Will add

- **Dependency**: Platform bare metal single-node support status assessed with the Bare Metal Networking team. If keepalived cannot be configured for single-node clusters, the limitation is documented and `platform: none` remains the only supported path

### Tech Preview -> GA

- Full test coverage including upgrades (y-stream and z-stream) on post-transition clusters
- SLOs documented and validated: target transition duration (SNO → HA compact), success rate threshold, and maximum time in the 2-member etcd window
- Monitoring and telemetry for transition metrics: Prometheus metrics exposed (transition_started, transition_completed, transition_failed, transition_duration_seconds) with alerts defined for stuck transitions exceeding SLO thresholds
- Support procedures documented
- Feature gate moved to `Default` feature set

> - A plan is defined for improving the UX around optional (OLM-managed) operators during topology transitions; actually managing them remains out of scope for this enhancement and is left for a future enhancement (see [Optional (OLM-Managed) Operators and Topology Changes](#optional-olm-managed-operators-and-topology-changes))

Reviewer 9: We should mention the following tests here to some extent, without needing to dig down to find out that the plan still have to be written.

> Before this feature reaches GA, a plan (not necessarily a full implementation) for improving the UX around optional operators during a topology transition — e.g., surfacing which installed OLM operators have unknown or unsupported topology compatibility — must be defined. Actually detecting, blocking, or remediating transitions based on optional-operator topology support remains future work.

Also, it is not just a UX thing, this has functionality and maintanance implications. 

See https://github.com/openshift/enhancements/pull/2008#discussion_r3732315946 for more context.

Author: > Also, it is not just a UX thing, this has functionality and maintanance implications.

Will append the functionality/maintenance implications. 

> We should mention the following tests here to some extent

I'm not sure what tests you are referring to though. It's explicitly saying we want it for tech preview but it is future work

Reviewer 9: sorry for the typo: I meant to say the following _text_, not tests

6. The API server validates `controlPlaneTopology` against the `DesiredControlPlaneTopologyMode` enum, rejecting unsupported topology modes before accepting the write

##### During Transition

7. The topology transition controller in CCO detects the `controlPlaneTopology` change and validates preconditions:
   - Every ClusterOperator other than cluster-config-operator itself reports `Available=True`, `Progressing=False`, `Degraded=False`
   - Exactly 3 nodes with `node-role.kubernetes.io/control-plane` or `node-role.kubernetes.io/master` labels are present, all schedulable and `Ready`
   - No dedicated worker nodes are present (the initial implementation targets compact clusters only; clusters with dedicated workers require a different `infrastructureTopology` mapping that is not yet supported)
   - etcd already reports quorum, is not mid-scaling, and already has 3 voting members — i.e., step 2's node-driven scaling has already finished
   If any precondition fails — including an etcd that has not yet finished scaling — the controller does not admit the transition; it records the reason and re-evaluates on the next sync (see [Failure Handling](#failure-handling)).

> 8. Once preconditions pass, the controller sets `Upgradeable=False` and a `Progressing` condition on the CCO `ClusterOperator` in the same update, signaling that a transition is in progress and preventing CVO from initiating an upgrade

Reviewer 11: could there be a small window here where CVO would be able to start an upgrade ?

Author: Technically yes, prior to CCO reconciliation of the CLI change, admittedly a window but very tight.

Author: Updated to do a last-second upgrade check to make sure one hasn't initiated before setting upgradable false, also added a comment to the implementation PR so that I don't forget to add the actual code to do this.

- Learner instances are used before promoting members to minimize the promotion window
- No availability guarantee during transitions; administrators should treat scaling operations as a maintenance window
- If etcd scaling fails during the 2-member window, quorum is lost and manual recovery via `quorum-restore.sh` is required
- Future iterations may explore admitting two learners simultaneously and promoting only when both are ready, eliminating the 2-member voting window entirely, but that is out of scope for this enhancement

#### Risk: Transition Fails Partway Through

**Risk**: A transition may fail after some operators have begun reconfiguring but before the transition completes, leaving the cluster in an intermediate state. Examples of failures:

- **etcd quorum loss**: during the node-driven prerequisite scaling (before the controller admits the transition), etcd scales to 2 members, a network partition occurs between them, both lose quorum, and the API becomes unavailable. This requires manual recovery via `quorum-restore.sh`, independent of the topology transition controller.

> - **Node readiness**: a new control-plane node becomes `NotReady` during the transition (e.g., disk pressure, network misconfiguration), preventing etcd or static pods from starting.

Reviewer 11: it's partially mentioned here - but a related risk is a user joining a node of a different architecture (which is not supported for control plane)

Reviewer 4: That node wouldn't achieve ready state, so it would be caught by our pre-transition checks.

Defining an OLM manifest mechanism for operators to declare topology support (and any catalog requirements built on it) would let the transition controller and catalog tooling reason about optional-operator compatibility with confidence instead of assuming it.
Designing and implementing that mechanism, and any transition-blocking behavior based on it, is out of scope for this enhancement and is left for a dedicated future enhancement — see [Open Questions](#open-questions).

Before this feature reaches GA, a plan (not necessarily a full implementation) for improving the UX around optional operators during a topology transition — e.g., surfacing which installed OLM operators have unknown or unsupported topology compatibility — must be defined. Actually detecting, blocking, or remediating transitions based on optional-operator topology support remains future work.

#### One-Way Transitions (Initially)

> The initial implementation supports only SNO → HA compact. Reverse transitions (HA → SNO) and other paths are future work. Administrators who transition cannot revert without redeploying.

Reviewer 11: will this be prevented through the CLI?

Author: It is currently prevented in the API (only valid values are allowed), at the CLI level and in the CCO controller. Those mechanisms are defined in various places in the doc. Currently the CLI and CCO controller duplicate logic that we will be merging into a new machanism as part of working towards dev and tech preview.

Author: clarified the statement with regards to prevention
